# Reaching Beyond the Stars In Recommending Thai Restaurants in Las Vegas: A Sentiment Detecting Approach to Rating Reviews as a Complement to User Ratings

## ABSTRACT

In line with career prospects in social networking applications, this classification project deals with recommender systems and focuses on the problem of matching a user to a new item. Based on the business and review academic datasets from the 2015 Yelp Dataset Challenge, the task is to classify a Thai restaurant in Las Vegas that is new to a user as a restaurant for the user to experience or not. Performance is measured, as a matter of classification accuracy, by a restaurant rating that is predicted correctly for a user from user ratings and from lexicon-extracted sentiment scores in review text. The merged working dataset has 990,627 reviews, of which 405,760 target one of the 4,960 restaurants in Las Vegas. In line with the “Yelp Restaurant Lexicon” (Kiritchenko, Zhu, Cherry, & Mohammad, 2014b) whose units of analysis are reviews of Yelp Restaurants in Phoenix (AZ), we narrow down the reviews to reviewers of restaurants in Phoenix, although not exclusively. We further narrow the reviews to Thai restaurants to accommodate R and RAM limits. Using restaurant ratings given by reviewers and sentiment scores, we select a user-based collaborative approach to predict missing ratings and to deliver to the user top recommendations. 

## CODE AND RESULTS FOR THE SENTIMENT DETECTION EXERCISE

The objectives of this kernel are to extract sentiment scores from Thai restaurant reviews using lexicons and to perform a preliminary comparison assessment of the review scores and the ratings of the restaurants given by individual reviewers.

In this kernel, we rely on a working dataset that was created in a previous phase of the project from the merger of a slightly transformed version of the Yelp_academic_dataset_business to the Yelp_academic_dataset_review (key: business_id, method: left, left dataset: Yelp_academic_dataset_review). For lexicon compatibility reasons and for RAM limitation reasons, we filtered the working dataset respectively by users who have rated restaurants in Phoenix (AZ) and by the "Thai" category of restaurants. Before filtering the working dataset by Thai restaurants, we combined review text by unique users and averaged review_ratings by unique users. The restaurant review data files in this kernel were derived from this working dataset. The lexicon data files were downloaded from the internet.

In the next few code cells, we will be reading four files: 

1- ThaiTextByUniqueUserBiz.csv, which has three separate columns of interest: user_id, business_id and text 

2- Yelp-restaurant-reviews-AFFLEX-NEGLEX-unigrams.txt, which was downloaded from http://saifmohammad.com/Lexicons/Yelp-restaurant-reviews.zip

3- AFINN-emoticons-en-165.txt, which is a lexicon derived from the content of the AFINN_emoticon-8 downloaded from https://github.com/fnielsen/afinn/blob/master/afinn/data/AFINN-emoticon-8.txt and manually copied to the AFINN-en-165 lexicon downloaded from https://github.com/fnielsen/afinn/blob/master/afinn/data/AFINN-en-165.txt

4- ThaiReviewRatingsByUserBiz.pkl, which has three columns of interest: user_id, business_id and review_ratings.

The files in items 1 and 4 were derived from the same working dataset. We made a decision to re-merge the two with the keys 'user_id' and 'business_id' at the end of this kernel to prevent confusion and to compare the various sentiment scores with the review_ratings.

### Import Utility Packages

We acknowledge that some of these packages are overkill. You may wish to select what you need to decrease the burden on your memory.

In [1]:
import sys
import re
import os
import shutil
import commands
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import itertools
import nltk
nltk.download('all')

[nltk_data] Downloading collection u'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to
[nltk_data]    |     C:\Users\Luc\AppData\Roaming\nltk_data...
[nltk_data]    |   Package abc is already up-to-date!
[nltk_data]    | Downloading package alpino to
[nltk_data]    |     C:\Users\Luc\AppData\Roaming\nltk_data...
[nltk_data]    |   Package alpino is already up-to-date!
[nltk_data]    | Downloading package biocreative_ppi to
[nltk_data]    |     C:\Users\Luc\AppData\Roaming\nltk_data...
[nltk_data]    |   Package biocreative_ppi is already up-to-date!
[nltk_data]    | Downloading package brown to
[nltk_data]    |     C:\Users\Luc\AppData\Roaming\nltk_data...
[nltk_data]    |   Package brown is already up-to-date!
[nltk_data]    | Downloading package brown_tei to
[nltk_data]    |     C:\Users\Luc\AppData\Roaming\nltk_data...
[nltk_data]    |   Package brown_tei is already up-to-date!
[nltk_data]    | Downloading package cess_cat to
[nltk_data]    |     C:\Users\Luc\AppDat

True

### Read the Thai Restaurant Review File

We choose a csv file rather than a pkl file and encode with utf-8 in order to avoid complications with accented charaters, such as 'ê' in crêpe.  

In [2]:
ThaiTextByUniqueUserBiz = pd.read_csv("ThaiTextByUniqueUserBiz.csv", encoding='utf-8')
ThaiTextByUniqueUserBiz

Unnamed: 0                 user_id             business_id  \
0               0  --65q1FpAL_UQtVZ2PTGew  JiLK9QPjd53pOBEAaY83lw   
1               1  --ijvARuRJhZrBdS9_jF2A  ApUCpJ9aa6yVgsde16gYrg   
2               2  --ohLoec6PU9_yxhbIlVWg  2XXwiASSS6685OhWWnIt_A   
3               3  --qEXbk-cA0HmbPyhcffdA  CVos739DJ06t8-dNiRMyeQ   
4               4  --qEXbk-cA0HmbPyhcffdA  jQST5lkLGX9L52-A10TGTQ   
5               5  -0fMBkX7QvWKQrtOp7H-GQ  3rqoxOasrRKxNubxjLSElA   
6               6  -2EuoueswhqEERWezJY8gw  cInzGnaFZ3EIItvFXl1MvQ   
7               7  -2Ig3GSBkj8JQT8eETmDPg  d-YNxMKL6ZhkiRhfUPxKHg   
8               8  -3WzrbWjnaKg2QWAsouy_g  jQST5lkLGX9L52-A10TGTQ   
9               9  -45GJdo8Ye8A1AStuUZp9Q  -SNpLwJNup8N96yq7sBJyw   
10             10  -4c_mgQdLH5axJ3j2In5_Q  WPmamMTGAmNYXGoXW1mWyQ   
11             11  -7R1u0HzHKmhLy9qE2MBpw  KTF-E3NfkJy2wiwcgOPyVQ   
12             12  -8BqfYouq3o_UoazAQWwNw  a_wK-2KhPu-8DAwwRObr8A   
13             13  -8BqfYouq3o_UoazAQWwNw  jQST5lkLGX9L52-A10TGTQ   
14             14  -8gRkiYaVm3zfoQ4pcg75w  UxiSHVZxMdey7vRwm1fQyA   
15             15  -8pbvWZH7Czk9YW1UkW4Ng  MDtjD14H1sGLc4tSg0sUhw   
16             16  -9-fkZ72_Qg4E6YYYXMqSg  4nnMgD9X62YrMqkQKhx-Pg   
17             17  -9g6w1xoj6-4iZH29P3h7g  kGEW4XXJQ2FS94gZv_N7VA   
18             18  -A01aSKVuOm42FnhvOCdKA  JiLK9QPjd53pOBEAaY83lw   
19             19  -ARd7byPUILfnFVlKcn0Yg  wct7rZKyZqZftzmAU-vhWQ   
20             20  -BQFGG_hrORLkEs8oigCjg  qcm7pfIdNn9XBuPEtoogbw   
21             21  -BVv1TDLLphHzgKw-eAJJQ  fJzKYljToXOauSohw9cMIA   
22             22  -BVv1TDLLphHzgKw-eAJJQ  puFrm8eNizztqaWr_e32pQ   
23             23  -BrtOvg4tL1xcaQnQTaZow  oXQmAzFj_qKNhUGYGNWLSA   
24             24  -CyBQG3dc4UnpluY7UdMOA  ujLZmyy11g1JHCQTxRA3Dw   
25             25  -DRza4wuHHWfQx5HcG6qaw  NCtzWkMbE13r2M2Sg0wH9w   
26             26  -E7e4sTuVAHwwWjQYBG07w  5W48_DnrXVD7EbtmE4pxOQ   
27             27  -EFuxDYchSSVkb4Q9Iivpg  shCdCHRbnY5FTMJbWl-myQ   
28             28  -EyEj5BujVFisco6OwmR8A  puFrm8eNizztqaWr_e32pQ   
29             29  -F32Vl8Rk4dwsmk0f2wRIw  NH67MdKaFGNcP-dlu56pyw   
...           ...                     ...                     ...   
10881       10881  zkhOTlhe6dn-jrwDpYDN6Q  LbBxrQJl-ny02-eCM1LYNg   
10882       10882  zlcHQII8dyI8I0LHGj8nOA  90AXjqb4O-wrTHDKDoDUzg   
10883       10883  zlcHQII8dyI8I0LHGj8nOA  AsX-6ECbV83zGJLUVMre9w   
10884       10884  zlneJ82kppmQXOUGHqCLaQ  RlfX4muX5LfJsvmI9qWGvw   
10885       10885  zmZtT1T6-J4NcqP8j1L5jA  Nz_AasmpsQ8MLSqhCTRVoA   
10886       10886  zmZtT1T6-J4NcqP8j1L5jA  joxWCp6dgN-kTE9GMziwjA   
10887       10887  zn81QpflLDUaGZkCMUowCg  8qrICL2tS2Rq7b5gxUdQwQ   
10888       10888  zn81QpflLDUaGZkCMUowCg  I1rvqU2k5UQGo2lGdY6hyw   
10889       10889  zn81QpflLDUaGZkCMUowCg  NH67MdKaFGNcP-dlu56pyw   
10890       10890  zn81QpflLDUaGZkCMUowCg  qyNtVViurIcChc35mfYIEw   
10891       10891  zn81QpflLDUaGZkCMUowCg  vtQOervVVTXjhvSZQiZ6PA   
10892       10892  znDOmt2ifMXWiAkrhjiuig  AsX-6ECbV83zGJLUVMre9w   
10893       10893  zp-DF3qfvOn5ko_vjpQLOg  KPoTixdjoJxSqRSEApSAGg   
10894       10894  zqHznU4iL06NziZIEGWHJw  lliksv-tglfUz1T3B3vgvA   
10895       10895  zrBmkDDLS94GYexyp0LyqQ  NH67MdKaFGNcP-dlu56pyw   
10896       10896  zrO1ENicvYdPsQk8ykJOkg  a1t31qMLd5fQocEjbSJ61A   
10897       10897  zs1msKnmTFD3iV2u69USuA  PXmR1MgOAWB066XH20HjxQ   
10898       10898  zs6wQGh1r726ZzaNKRa-bw  JiLK9QPjd53pOBEAaY83lw   
10899       10899  zs6wQGh1r726ZzaNKRa-bw  j2a5uJz76rK9uTRgLn5TdQ   
10900       10900  zs6wQGh1r726ZzaNKRa-bw  xcxkEmy4CD-qaJUqprvpHA   
10901       10901  zssolmdLpaX1tpRRMDWYwg  0udEgNqy5rLR5pZ4kD19Og   
10902       10902  zulNp3NWnv7sYODNZ1Xrow  o15PeOAUzpcCl8ngk0lMHw   
10903       10903  zuoeE7GdXXlCgr995ImWfQ  xcxkEmy4CD-qaJUqprvpHA   
10904       10904  zv4i7JjhI9v9j4ZzX7TGDw  NCtzWkMbE13r2M2Sg0wH9w   
10905       10905  zvFDYEFo_xO8VqLQfmB-DA  shCdCHRbnY5FTMJbWl-myQ   
10906       10906  zw5NmE_epbvJ22xOYLdIoQ  cBwc3dhdHw0emmg9nd5SXw   
10907

In [3]:
ThaiTextByUniqueUserBiz.drop(ThaiTextByUniqueUserBiz.columns[[0]],axis=1, inplace=True)
ThaiTextByUniqueUserBiz

user_id             business_id  \
0      --65q1FpAL_UQtVZ2PTGew  JiLK9QPjd53pOBEAaY83lw   
1      --ijvARuRJhZrBdS9_jF2A  ApUCpJ9aa6yVgsde16gYrg   
2      --ohLoec6PU9_yxhbIlVWg  2XXwiASSS6685OhWWnIt_A   
3      --qEXbk-cA0HmbPyhcffdA  CVos739DJ06t8-dNiRMyeQ   
4      --qEXbk-cA0HmbPyhcffdA  jQST5lkLGX9L52-A10TGTQ   
5      -0fMBkX7QvWKQrtOp7H-GQ  3rqoxOasrRKxNubxjLSElA   
6      -2EuoueswhqEERWezJY8gw  cInzGnaFZ3EIItvFXl1MvQ   
7      -2Ig3GSBkj8JQT8eETmDPg  d-YNxMKL6ZhkiRhfUPxKHg   
8      -3WzrbWjnaKg2QWAsouy_g  jQST5lkLGX9L52-A10TGTQ   
9      -45GJdo8Ye8A1AStuUZp9Q  -SNpLwJNup8N96yq7sBJyw   
10     -4c_mgQdLH5axJ3j2In5_Q  WPmamMTGAmNYXGoXW1mWyQ   
11     -7R1u0HzHKmhLy9qE2MBpw  KTF-E3NfkJy2wiwcgOPyVQ   
12     -8BqfYouq3o_UoazAQWwNw  a_wK-2KhPu-8DAwwRObr8A   
13     -8BqfYouq3o_UoazAQWwNw  jQST5lkLGX9L52-A10TGTQ   
14     -8gRkiYaVm3zfoQ4pcg75w  UxiSHVZxMdey7vRwm1fQyA   
15     -8pbvWZH7Czk9YW1UkW4Ng  MDtjD14H1sGLc4tSg0sUhw   
16     -9-fkZ72_Qg4E6YYYXMqSg  4nnMgD9X62YrMqkQKhx-Pg   
17     -9g6w1xoj6-4iZH29P3h7g  kGEW4XXJQ2FS94gZv_N7VA   
18     -A01aSKVuOm42FnhvOCdKA  JiLK9QPjd53pOBEAaY83lw   
19     -ARd7byPUILfnFVlKcn0Yg  wct7rZKyZqZftzmAU-vhWQ   
20     -BQFGG_hrORLkEs8oigCjg  qcm7pfIdNn9XBuPEtoogbw   
21     -BVv1TDLLphHzgKw-eAJJQ  fJzKYljToXOauSohw9cMIA   
22     -BVv1TDLLphHzgKw-eAJJQ  puFrm8eNizztqaWr_e32pQ   
23     -BrtOvg4tL1xcaQnQTaZow  oXQmAzFj_qKNhUGYGNWLSA   
24     -CyBQG3dc4UnpluY7UdMOA  ujLZmyy11g1JHCQTxRA3Dw   
25     -DRza4wuHHWfQx5HcG6qaw  NCtzWkMbE13r2M2Sg0wH9w   
26     -E7e4sTuVAHwwWjQYBG07w  5W48_DnrXVD7EbtmE4pxOQ   
27     -EFuxDYchSSVkb4Q9Iivpg  shCdCHRbnY5FTMJbWl-myQ   
28     -EyEj5BujVFisco6OwmR8A  puFrm8eNizztqaWr_e32pQ   
29     -F32Vl8Rk4dwsmk0f2wRIw  NH67MdKaFGNcP-dlu56pyw   
...                       ...                     ...   
10881  zkhOTlhe6dn-jrwDpYDN6Q  LbBxrQJl-ny02-eCM1LYNg   
10882  zlcHQII8dyI8I0LHGj8nOA  90AXjqb4O-wrTHDKDoDUzg   
10883  zlcHQII8dyI8I0LHGj8nOA  AsX-6ECbV83zGJLUVMre9w   
10884  zlneJ82kppmQXOUGHqCLaQ  RlfX4muX5LfJsvmI9qWGvw   
10885  zmZtT1T6-J4NcqP8j1L5jA  Nz_AasmpsQ8MLSqhCTRVoA   
10886  zmZtT1T6-J4NcqP8j1L5jA  joxWCp6dgN-kTE9GMziwjA   
10887  zn81QpflLDUaGZkCMUowCg  8qrICL2tS2Rq7b5gxUdQwQ   
10888  zn81QpflLDUaGZkCMUowCg  I1rvqU2k5UQGo2lGdY6hyw   
10889  zn81QpflLDUaGZkCMUowCg  NH67MdKaFGNcP-dlu56pyw   
10890  zn81QpflLDUaGZkCMUowCg  qyNtVViurIcChc35mfYIEw   
10891  zn81QpflLDUaGZkCMUowCg  vtQOervVVTXjhvSZQiZ6PA   
10892  znDOmt2ifMXWiAkrhjiuig  AsX-6ECbV83zGJLUVMre9w   
10893  zp-DF3qfvOn5ko_vjpQLOg  KPoTixdjoJxSqRSEApSAGg   
10894  zqHznU4iL06NziZIEGWHJw  lliksv-tglfUz1T3B3vgvA   
10895  zrBmkDDLS94GYexyp0LyqQ  NH67MdKaFGNcP-dlu56pyw   
10896  zrO1ENicvYdPsQk8ykJOkg  a1t31qMLd5fQocEjbSJ61A   
10897  zs1msKnmTFD3iV2u69USuA  PXmR1MgOAWB066XH20HjxQ   
10898  zs6wQGh1r726ZzaNKRa-bw  JiLK9QPjd53pOBEAaY83lw   
10899  zs6wQGh1r726ZzaNKRa-bw  j2a5uJz76rK9uTRgLn5TdQ   
10900  zs6wQGh1r726ZzaNKRa-bw  xcxkEmy4CD-qaJUqprvpHA   
10901  zssolmdLpaX1tpRRMDWYwg  0udEgNqy5rLR5pZ4kD19Og   
10902  zulNp3NWnv7sYODNZ1Xrow  o15PeOAUzpcCl8ngk0lMHw   
10903  zuoeE7GdXXlCgr995ImWfQ  xcxkEmy4CD-qaJUqprvpHA   
10904  zv4i7JjhI9v9j4ZzX7TGDw  NCtzWkMbE13r2M2Sg0wH9w   
10905  zvFDYEFo_xO8VqLQfmB-DA  shCdCHRbnY5FTMJbWl-myQ   
10906  zw5NmE_epbvJ22xOYLdIoQ  cBwc3dhdHw0emmg9nd5SXw   
10907  zxQaAt4awDFVWme2I9mFgg  0udEgNqy5rLR5pZ4kD19Og   
10908  zxcrlC3cmH5S2TGIxuLwBw  MDtjD14H1sGLc4tSg0sUhw   
10909  zxcrlC3cmH5S2TGIxuLwBw  kGEW4XXJQ2FS94gZv_N7VA   
10910  zyor9BbfHNjTsaRFfePRwQ  apGVTRZRCQ9-89hu2qW-vw   

                                                    text  
0      I'm a big fan of this place and have dropped i...  
1      Food was ok but the service was less than exce...  
2      I got the Penang curry and have to say the foo...  
3      To sum up in one sentence: "I only go to Thai ...  
4      I LOVE THIS PLACE!\r\r\n\r\r\nIt's a cute mom-...  
5      The food was delicious and the service was ama...  
6      My Girlfriend and I eat here occasionally and ...  
7      Very fri

### Read the Yelp Restaurant Review Lexicon File


Although a bigram-based lexicon is available from the Yelp Restaurant Review Lexicon website file, we choose for this exercise the unigram-based lexicon for ease of comparison with the AFINN lexicon. We first assign column names to the four types of values found in the unigram-based text file although we will be using for this exercise the 'word' and 'score' values only.

In [4]:
colnames = ['word', 'score', 'pos', 'neg']
YelpSentiment = pd.read_table("Yelp-restaurant-reviews-AFFLEX-NEGLEX-unigrams.txt", names=colnames)
YelpSentiment

word  score  pos  neg
0         overpowering_NEGFIRST  3.798  139    0
1                          yumm  3.716  128    0
2                         faves  3.306  256    2
3                        yummmm  3.250   80    0
4                     satisfies  3.238   79    0
5          disappoints_NEGFIRST  3.100  208    2
6                       bosnian  3.061   66    0
7                      combines  3.046   65    0
8                      vinegars  2.967   60    0
9                     exquisite  2.955  240    3
10              regret_NEGFIRST  2.953  360    5
11                        yummm  2.942  118    1
12                   seasonally  2.899   56    0
13                 approachable  2.881   55    0
14                     heavenly  2.875  667   11
15                        yummo  2.863   54    0
16                       payton  2.863   54    0
17                       awsome  2.863  109    1
18                        pablo  2.845   53    0
19                          a++  2.845   53    0
20                     penzey's  2.845   53    0
21                  awesomeness  2.833  159    2
22                       camper  2.826  105    1
23                       cutest  2.807  103    1
24          disappoint_NEGFIRST  2.802  982   18
25                      burmese  2.768   49    0
26                    popsicles  2.748   48    0
27                       godiva  2.727   47    0
28               pushy_NEGFIRST  2.727   47    0
29                    delicioso  2.695   92    1
...                         ...    ...  ...  ...
38352                        ym -3.542    0   10
38353            reply_NEGFIRST -3.542    0   10
38354                  dirtiest -3.542    1   21
38355                 cockroach -3.568    6   78
38356                  unedible -3.586    1   22
38357                    rudest -3.601    2   34
38358                 uneatable -3.629    0   11
38359                   insipid -3.629    0   11
38360                  blandest -3.709    1   25
38361               inexcusable -3.709    2   38
38362             argumentative -3.709    0   12
38363              insulted_NEG -3.709    0   12
38364           return_NEGFIRST -3.719   43  577
38365              inedible_NEG -3.783    1   27
38366                     oggis -3.783    0   13
38367        redeeming_NEGFIRST -3.783    0   13
38368         greeting_NEGFIRST -3.852    1   29
38369          compensation_NEG -3.852    0   14
38370          apology_NEGFIRST -3.852    6  104
38371            stars_NEGFIRST -3.878    4   76
38372                    rudely -3.960    6  116
38373            inconvenienced -3.977    0   16
38374       apologized_NEGFIRST -3.977    0   16
38375        patronize_NEGFIRST -4.034    0   17
38376           pepper_NEGFIRST -4.034    0   17
38377  acknowledgement_NEGFIRST -4.034    0   17
38378                  grossest -4.088    0   18
38379               apology_NEG -4.114    3   77
38380           refund_NEGFIRST -4.140    0   19
38381        returning_NEGFIRST -4.440    0   26

[38382 rows x 4 columns]

### Filter out the NEGLEX elements from the Yelp Restaurant Review Lexicon  

The unigram lexicon for Yelp restaurant reviews has two scales, each with words (common or not with the other scale) and its own sentiment scores. For the purpose of this excercise, we filter out the NEGLEX scale as it provides positive and negative values for some words with negative connotations. We are seeking alignment with the AFINN lexicons for comparison purposes. The NEGFLEX lexicon is provided for reference purposes only.

In [5]:
YelpSentimentAFFLEX = YelpSentiment[YelpSentiment.word.str.contains("_NEG") == False]
YelpSentimentAFFLEX = YelpSentimentAFFLEX.reset_index(drop=True)
YelpSentimentAFFLEX

word  score  pos  neg
0                yumm  3.716  128    0
1               faves  3.306  256    2
2              yummmm  3.250   80    0
3           satisfies  3.238   79    0
4             bosnian  3.061   66    0
5            combines  3.046   65    0
6            vinegars  2.967   60    0
7           exquisite  2.955  240    3
8               yummm  2.942  118    1
9          seasonally  2.899   56    0
10       approachable  2.881   55    0
11           heavenly  2.875  667   11
12              yummo  2.863   54    0
13             payton  2.863   54    0
14             awsome  2.863  109    1
15              pablo  2.845   53    0
16                a++  2.845   53    0
17           penzey's  2.845   53    0
18        awesomeness  2.833  159    2
19             camper  2.826  105    1
20             cutest  2.807  103    1
21            burmese  2.768   49    0
22          popsicles  2.748   48    0
23             godiva  2.727   47    0
24          delicioso  2.695   92    1
25            chino's  2.663   44    0
26               joel  2.663   44    0
27               timo  2.663   44    0
28            barbara  2.663   44    0
29               maui  2.651  177    3
...               ...    ...  ...  ...
29666        promoter -3.341    0    8
29667     unflavorful -3.341    0    8
29668    unacceptably -3.341    0    8
29669           xooro -3.341    0    8
29670    unacceptable -3.346   20  189
29671      imposition -3.447    0    9
29672            eggo -3.447    0    9
29673           livid -3.447    1   19
29674          curtly -3.447    0    9
29675          mildew -3.447    0    9
29676          spoonz -3.447    0    9
29677      over-rated -3.479    2   30
29678            tock -3.495    1   20
29679   disrespectful -3.511    2   31
29680        defended -3.542    0   10
29681        vomiting -3.542    2   32
29682              ym -3.542    0   10
29683        dirtiest -3.542    1   21
29684       cockroach -3.568    6   78
29685        unedible -3.586    1   22
29686          rudest -3.601    2   34
29687       uneatable -3.629    0   11
29688         insipid -3.629    0   11
29689        blandest -3.709    1   25
29690     inexcusable -3.709    2   38
29691   argumentative -3.709    0   12
29692           oggis -3.783    0   13
29693          rudely -3.960    6  116
29694  inconvenienced -3.977    0   16
29695        grossest -4.088    0   18

[29696 rows x 4 columns]

In [6]:
YelpSentimentNEGFLEX = YelpSentiment[YelpSentiment.word.str.contains("_NEG") == True]
YelpSentimentNEGFLEX

word  score  pos  neg
0         overpowering_NEGFIRST  3.798  139    0
5          disappoints_NEGFIRST  3.100  208    2
10              regret_NEGFIRST  2.953  360    5
24          disappoint_NEGFIRST  2.802  982   18
28               pushy_NEGFIRST  2.727   47    0
30         pretentious_NEGFIRST  2.692  138    2
31           overpower_NEGFIRST  2.685   45    0
38               affordable_NEG  2.640   43    0
46              rushed_NEGFIRST  2.606   84    1
51            skimping_NEGFIRST  2.570   40    0
62        disappointed_NEGFIRST  2.497  990   25
63              failed_NEGFIRST  2.494   37    0
78                beat_NEGFIRST  2.432  821   22
81             detract_NEGFIRST  2.411   34    0
89                     tour_NEG  2.382   33    0
96              stuffy_NEGFIRST  2.368   66    1
106                     gem_NEG  2.330  128    3
132       dissapointed_NEGFIRST  2.257   29    0
136                devoured_NEG  2.223   28    0
148           pressure_NEGFIRST  2.223   28    0
152           overdone_NEGFIRST  2.206   56    1
154           hesitate_NEGFIRST  2.193  196    6
176             skimpy_NEGFIRST  2.152   26    0
178             pricey_NEGFIRST  2.152   26    0
179                  carrot_NEG  2.152   26    0
198             closer_NEGFIRST  2.114   25    0
207                  cartel_NEG  2.114   25    0
211             hassle_NEGFIRST  2.114   25    0
214                  shaved_NEG  2.114   25    0
230                 fastest_NEG  2.075   24    0
...                         ...    ...  ...  ...
38322      acknowledge_NEGFIRST -3.341    2   26
38326             precooked_NEG -3.341    0    8
38328               voucher_NEG -3.341    1   17
38329          disrespected_NEG -3.341    0    8
38331         response_NEGFIRST -3.369    3   36
38332           spices_NEGFIRST -3.395    1   18
38333              eye_NEGFIRST -3.395    1   18
38334           flavor_NEGFIRST -3.410   54  529
38338       compensate_NEGFIRST -3.447    0    9
38339           comped_NEGFIRST -3.447    0    9
38344            again_NEGFIRST -3.483   26  279
38345        seasoning_NEGFIRST -3.495    5   62
38349          manager_NEGFIRST -3.542    2   32
38351   acknowledgment_NEGFIRST -3.542    0   10
38353            reply_NEGFIRST -3.542    0   10
38363              insulted_NEG -3.709    0   12
38364           return_NEGFIRST -3.719   43  577
38365              inedible_NEG -3.783    1   27
38367        redeeming_NEGFIRST -3.783    0   13
38368         greeting_NEGFIRST -3.852    1   29
38369          compensation_NEG -3.852    0   14
38370          apology_NEGFIRST -3.852    6  104
38371            stars_NEGFIRST -3.878    4   76
38374       apologized_NEGFIRST -3.977    0   16
38375        patronize_NEGFIRST -4.034    0   17
38376           pepper_NEGFIRST -4.034    0   17
38377  acknowledgement_NEGFIRST -4.034    0   17
38379               apology_NEG -4.114    3   77
38380           refund_NEGFIRST -4.140    0   19
38381        returning_NEGFIRST -4.440    0   26

[8685 rows x 4 columns]

### Create Two Python Dictionaries with Sentiment Scores for the Sentiment Detection Algorithm

In this section, the objective of the code cells are (1) to identify the dictionary format used with the AFINN lexicon for the sentiment dectection algorithm and (2) the creation of a Python dictionary in the same format with the Yelp Restaurant Review Lexicon from the YelpSentimentAFFLEX dataframe. The Python dictionaries will be used in sentiment detection algorithms in the next section. 

#### Create the AFINN dictionary

In [7]:
sentiment_dictionary = {}
for line in open('AFINN-emoticons-en-165.txt'):
    word, score = line.split('\t')
    sentiment_dictionary[word] = int(score)
sentiment_dictionary 

{'unimaginative': -2,
 'limited': -1,
 'unscientific': -2,
 'suicidal': -2,
 'pardon': 2,
 'desirable': 2,
 'foul': -3,
 'obstruction': -2,
 'protest': -2,
 'lurking': -1,
 'controversial': -2,
 'hating': -3,
 'ridiculous': -3,
 'hate': -3,
 '\\o/': 3,
 'aggression': -2,
 'poorly': -2,
 'stinks': -2,
 'infuriates': -2,
 'regretted': -2,
 'violate': -2,
 'granting': 1,
 'attracted': 1,
 'tremors': -2,
 'stinky': -2,
 'poorest': -2,
 'disability': -2,
 'condemns': -2,
 'sorry': -1,
 'regrets': -2,
 'struck': -1,
 'misreporting': -2,
 'compassion': 2,
 'misreports': -2,
 'hilarious': 2,
 'lurk': -1,
 'misunderstanding': -2,
 'distort': -2,
 'lololol': 4,
 'stolen': -2,
 'gratification': 2,
 'uncertain': -1,
 'stabbed': -2,
 'screaming': -2,
 'courageous': 2,
 'disturb': -2,
 'exaggerate': -2,
 'harried': -2,
 'solution': 1,
 'nigger': -5,
 'honor': 2,
 'pardons': 2,
 'delightfully': 3,
 'monopolized': -2,
 'illiteracy': -2,
 'triumph': 4,
 'enjoy': 2,
 'shithead': -4,
 'diverting': -1,
 '

#### Create the Yelp Sentiment dictionary

In [8]:
Yelp_sentiment_dictionary = YelpSentimentAFFLEX.set_index('word')['score'].to_dict()
Yelp_sentiment_dictionary

{'gai': 1.6890000000000001,
 'mid-week': 0.16800000000000001,
 'woods': 0.90200000000000002,
 'hanging': 0.13300000000000001,
 'woody': 0.59899999999999998,
 'northsight': -0.33299999999999996,
 'comically': -0.67400000000000004,
 'frou-frou': 0.35999999999999999,
 'cake-': 0.871,
 'originality': -0.098000000000000004,
 'calpico': 1.0529999999999999,
 'fattiness': 0.24199999999999999,
 'rawhide': -0.45100000000000001,
 'bringing': 0.089999999999999997,
 'tcby': -0.16300000000000001,
 'revelers': -0.45100000000000001,
 'caramels': 0.89300000000000002,
 'grueling': 0.155,
 'broiler': -0.90300000000000002,
 'caramely': 0.80200000000000005,
 'condessa': -0.29699999999999999,
 'wednesday': 0.56799999999999995,
 'broiled': 0.54600000000000004,
 'crotch': -0.044999999999999998,
 'stereotypical': -0.64800000000000002,
 'caramelo': 1.159,
 'bbqs': 0.35999999999999999,
 'chimichuri': 0.24199999999999999,
 "roscoe's": 0.92299999999999993,
 "tom's": -0.33799999999999997,
 'scrapes': 0.648000000000

#### Observations

We note that the two dictionaries have emoticons, that their word elements are lower-cased and that verbs are conjugated in various tenses. We also note that the Yelp Sentiment dictionary has both correctly spelled and mispelled words.

### Run the Sentiment Detection Algorithm with Dictionaries and NLTK's TweetTokenizer

As the restaurant reviews in our dataset has emoticons, we choose NLTK's TweetTokenizer to tokenize words from restaurant reviews with a view not to sacrifice the emoticons that provide key review sentiments. We acknowledge however that sarcasm may be difficult to detect with a simple dictionary method. We also choose to lower the case of all words and emoticons as both dictionaries have exclusively lower-case words and as the AFINN dictionary has both higher-case emoticons and their lower-case equivalents. We choose not to combine all the sentiment score summing lines in the same algorithm for the purposes of comprehension. We acknowledge that a combination of these summing lines in the same algorithm may save significant computational ressources with larger datasets.

#### Extract Sentiment Scores from Reviews Using Algorithm, TweetTokenizer and Yelp Restaurant Review Sentiment Dictionary

In [9]:
from nltk.tokenize import TweetTokenizer
tknzr = TweetTokenizer()
for row in range(len(ThaiTextByUniqueUserBiz)):
    n = ThaiTextByUniqueUserBiz.loc[row, 'text']
    words = tknzr.tokenize(n.lower())
    ThaiTextByUniqueUserBiz.loc[row,"YRsentiment"] = sum(Yelp_sentiment_dictionary.get(word, 0) for word in words)
ThaiTextByUniqueUserBiz  

user_id             business_id  \
0      --65q1FpAL_UQtVZ2PTGew  JiLK9QPjd53pOBEAaY83lw   
1      --ijvARuRJhZrBdS9_jF2A  ApUCpJ9aa6yVgsde16gYrg   
2      --ohLoec6PU9_yxhbIlVWg  2XXwiASSS6685OhWWnIt_A   
3      --qEXbk-cA0HmbPyhcffdA  CVos739DJ06t8-dNiRMyeQ   
4      --qEXbk-cA0HmbPyhcffdA  jQST5lkLGX9L52-A10TGTQ   
5      -0fMBkX7QvWKQrtOp7H-GQ  3rqoxOasrRKxNubxjLSElA   
6      -2EuoueswhqEERWezJY8gw  cInzGnaFZ3EIItvFXl1MvQ   
7      -2Ig3GSBkj8JQT8eETmDPg  d-YNxMKL6ZhkiRhfUPxKHg   
8      -3WzrbWjnaKg2QWAsouy_g  jQST5lkLGX9L52-A10TGTQ   
9      -45GJdo8Ye8A1AStuUZp9Q  -SNpLwJNup8N96yq7sBJyw   
10     -4c_mgQdLH5axJ3j2In5_Q  WPmamMTGAmNYXGoXW1mWyQ   
11     -7R1u0HzHKmhLy9qE2MBpw  KTF-E3NfkJy2wiwcgOPyVQ   
12     -8BqfYouq3o_UoazAQWwNw  a_wK-2KhPu-8DAwwRObr8A   
13     -8BqfYouq3o_UoazAQWwNw  jQST5lkLGX9L52-A10TGTQ   
14     -8gRkiYaVm3zfoQ4pcg75w  UxiSHVZxMdey7vRwm1fQyA   
15     -8pbvWZH7Czk9YW1UkW4Ng  MDtjD14H1sGLc4tSg0sUhw   
16     -9-fkZ72_Qg4E6YYYXMqSg  4nnMgD9X62YrMqkQKhx-Pg   
17     -9g6w1xoj6-4iZH29P3h7g  kGEW4XXJQ2FS94gZv_N7VA   
18     -A01aSKVuOm42FnhvOCdKA  JiLK9QPjd53pOBEAaY83lw   
19     -ARd7byPUILfnFVlKcn0Yg  wct7rZKyZqZftzmAU-vhWQ   
20     -BQFGG_hrORLkEs8oigCjg  qcm7pfIdNn9XBuPEtoogbw   
21     -BVv1TDLLphHzgKw-eAJJQ  fJzKYljToXOauSohw9cMIA   
22     -BVv1TDLLphHzgKw-eAJJQ  puFrm8eNizztqaWr_e32pQ   
23     -BrtOvg4tL1xcaQnQTaZow  oXQmAzFj_qKNhUGYGNWLSA   
24     -CyBQG3dc4UnpluY7UdMOA  ujLZmyy11g1JHCQTxRA3Dw   
25     -DRza4wuHHWfQx5HcG6qaw  NCtzWkMbE13r2M2Sg0wH9w   
26     -E7e4sTuVAHwwWjQYBG07w  5W48_DnrXVD7EbtmE4pxOQ   
27     -EFuxDYchSSVkb4Q9Iivpg  shCdCHRbnY5FTMJbWl-myQ   
28     -EyEj5BujVFisco6OwmR8A  puFrm8eNizztqaWr_e32pQ   
29     -F32Vl8Rk4dwsmk0f2wRIw  NH67MdKaFGNcP-dlu56pyw   
...                       ...                     ...   
10881  zkhOTlhe6dn-jrwDpYDN6Q  LbBxrQJl-ny02-eCM1LYNg   
10882  zlcHQII8dyI8I0LHGj8nOA  90AXjqb4O-wrTHDKDoDUzg   
10883  zlcHQII8dyI8I0LHGj8nOA  AsX-6ECbV83zGJLUVMre9w   
10884  zlneJ82kppmQXOUGHqCLaQ  RlfX4muX5LfJsvmI9qWGvw   
10885  zmZtT1T6-J4NcqP8j1L5jA  Nz_AasmpsQ8MLSqhCTRVoA   
10886  zmZtT1T6-J4NcqP8j1L5jA  joxWCp6dgN-kTE9GMziwjA   
10887  zn81QpflLDUaGZkCMUowCg  8qrICL2tS2Rq7b5gxUdQwQ   
10888  zn81QpflLDUaGZkCMUowCg  I1rvqU2k5UQGo2lGdY6hyw   
10889  zn81QpflLDUaGZkCMUowCg  NH67MdKaFGNcP-dlu56pyw   
10890  zn81QpflLDUaGZkCMUowCg  qyNtVViurIcChc35mfYIEw   
10891  zn81QpflLDUaGZkCMUowCg  vtQOervVVTXjhvSZQiZ6PA   
10892  znDOmt2ifMXWiAkrhjiuig  AsX-6ECbV83zGJLUVMre9w   
10893  zp-DF3qfvOn5ko_vjpQLOg  KPoTixdjoJxSqRSEApSAGg   
10894  zqHznU4iL06NziZIEGWHJw  lliksv-tglfUz1T3B3vgvA   
10895  zrBmkDDLS94GYexyp0LyqQ  NH67MdKaFGNcP-dlu56pyw   
10896  zrO1ENicvYdPsQk8ykJOkg  a1t31qMLd5fQocEjbSJ61A   
10897  zs1msKnmTFD3iV2u69USuA  PXmR1MgOAWB066XH20HjxQ   
10898  zs6wQGh1r726ZzaNKRa-bw  JiLK9QPjd53pOBEAaY83lw   
10899  zs6wQGh1r726ZzaNKRa-bw  j2a5uJz76rK9uTRgLn5TdQ   
10900  zs6wQGh1r726ZzaNKRa-bw  xcxkEmy4CD-qaJUqprvpHA   
10901  zssolmdLpaX1tpRRMDWYwg  0udEgNqy5rLR5pZ4kD19Og   
10902  zulNp3NWnv7sYODNZ1Xrow  o15PeOAUzpcCl8ngk0lMHw   
10903  zuoeE7GdXXlCgr995ImWfQ  xcxkEmy4CD-qaJUqprvpHA   
10904  zv4i7JjhI9v9j4ZzX7TGDw  NCtzWkMbE13r2M2Sg0wH9w   
10905  zvFDYEFo_xO8VqLQfmB-DA  shCdCHRbnY5FTMJbWl-myQ   
10906  zw5NmE_epbvJ22xOYLdIoQ  cBwc3dhdHw0emmg9nd5SXw   
10907  zxQaAt4awDFVWme2I9mFgg  0udEgNqy5rLR5pZ4kD19Og   
10908  zxcrlC3cmH5S2TGIxuLwBw  MDtjD14H1sGLc4tSg0sUhw   
10909  zxcrlC3cmH5S2TGIxuLwBw  kGEW4XXJQ2FS94gZv_N7VA   
10910  zyor9BbfHNjTsaRFfePRwQ  apGVTRZRCQ9-89hu2qW-vw   

                                                    text  YRsentiment  
0      I'm a big fan of this place and have dropped i...       43.539  
1      Food was ok but the service was less than exce...       -9.641  
2      I got the Penang curry and have to say the foo...        0.105  
3      To sum up in one sentence: "I only go to Thai ...      -16.096  
4      I LOVE THIS PLACE!\r\r\n\r\r\nIt's a cute mom-...        5.802  
5      The food was delicious and the service was ama..

#### Convert Yelp Review Sentiment Scores to a 1-to-5 Scale Using the Round Function without Decimals

In [10]:
OldMax = max(ThaiTextByUniqueUserBiz['YRsentiment'])
OldMin = min(ThaiTextByUniqueUserBiz['YRsentiment'])
NewMax = 5
NewMin = 1
OldRange = (OldMax - OldMin)
NewRange = (NewMax - NewMin)
for row in range(len(ThaiTextByUniqueUserBiz)):
    n = ThaiTextByUniqueUserBiz.loc[row, 'YRsentiment']
    ThaiTextByUniqueUserBiz.loc[row,"YRSentScore"] = (((n - OldMin) * NewRange / OldRange) + NewMin).round(decimals=0, out=None)
ThaiTextByUniqueUserBiz  

user_id             business_id  \
0      --65q1FpAL_UQtVZ2PTGew  JiLK9QPjd53pOBEAaY83lw   
1      --ijvARuRJhZrBdS9_jF2A  ApUCpJ9aa6yVgsde16gYrg   
2      --ohLoec6PU9_yxhbIlVWg  2XXwiASSS6685OhWWnIt_A   
3      --qEXbk-cA0HmbPyhcffdA  CVos739DJ06t8-dNiRMyeQ   
4      --qEXbk-cA0HmbPyhcffdA  jQST5lkLGX9L52-A10TGTQ   
5      -0fMBkX7QvWKQrtOp7H-GQ  3rqoxOasrRKxNubxjLSElA   
6      -2EuoueswhqEERWezJY8gw  cInzGnaFZ3EIItvFXl1MvQ   
7      -2Ig3GSBkj8JQT8eETmDPg  d-YNxMKL6ZhkiRhfUPxKHg   
8      -3WzrbWjnaKg2QWAsouy_g  jQST5lkLGX9L52-A10TGTQ   
9      -45GJdo8Ye8A1AStuUZp9Q  -SNpLwJNup8N96yq7sBJyw   
10     -4c_mgQdLH5axJ3j2In5_Q  WPmamMTGAmNYXGoXW1mWyQ   
11     -7R1u0HzHKmhLy9qE2MBpw  KTF-E3NfkJy2wiwcgOPyVQ   
12     -8BqfYouq3o_UoazAQWwNw  a_wK-2KhPu-8DAwwRObr8A   
13     -8BqfYouq3o_UoazAQWwNw  jQST5lkLGX9L52-A10TGTQ   
14     -8gRkiYaVm3zfoQ4pcg75w  UxiSHVZxMdey7vRwm1fQyA   
15     -8pbvWZH7Czk9YW1UkW4Ng  MDtjD14H1sGLc4tSg0sUhw   
16     -9-fkZ72_Qg4E6YYYXMqSg  4nnMgD9X62YrMqkQKhx-Pg   
17     -9g6w1xoj6-4iZH29P3h7g  kGEW4XXJQ2FS94gZv_N7VA   
18     -A01aSKVuOm42FnhvOCdKA  JiLK9QPjd53pOBEAaY83lw   
19     -ARd7byPUILfnFVlKcn0Yg  wct7rZKyZqZftzmAU-vhWQ   
20     -BQFGG_hrORLkEs8oigCjg  qcm7pfIdNn9XBuPEtoogbw   
21     -BVv1TDLLphHzgKw-eAJJQ  fJzKYljToXOauSohw9cMIA   
22     -BVv1TDLLphHzgKw-eAJJQ  puFrm8eNizztqaWr_e32pQ   
23     -BrtOvg4tL1xcaQnQTaZow  oXQmAzFj_qKNhUGYGNWLSA   
24     -CyBQG3dc4UnpluY7UdMOA  ujLZmyy11g1JHCQTxRA3Dw   
25     -DRza4wuHHWfQx5HcG6qaw  NCtzWkMbE13r2M2Sg0wH9w   
26     -E7e4sTuVAHwwWjQYBG07w  5W48_DnrXVD7EbtmE4pxOQ   
27     -EFuxDYchSSVkb4Q9Iivpg  shCdCHRbnY5FTMJbWl-myQ   
28     -EyEj5BujVFisco6OwmR8A  puFrm8eNizztqaWr_e32pQ   
29     -F32Vl8Rk4dwsmk0f2wRIw  NH67MdKaFGNcP-dlu56pyw   
...                       ...                     ...   
10881  zkhOTlhe6dn-jrwDpYDN6Q  LbBxrQJl-ny02-eCM1LYNg   
10882  zlcHQII8dyI8I0LHGj8nOA  90AXjqb4O-wrTHDKDoDUzg   
10883  zlcHQII8dyI8I0LHGj8nOA  AsX-6ECbV83zGJLUVMre9w   
10884  zlneJ82kppmQXOUGHqCLaQ  RlfX4muX5LfJsvmI9qWGvw   
10885  zmZtT1T6-J4NcqP8j1L5jA  Nz_AasmpsQ8MLSqhCTRVoA   
10886  zmZtT1T6-J4NcqP8j1L5jA  joxWCp6dgN-kTE9GMziwjA   
10887  zn81QpflLDUaGZkCMUowCg  8qrICL2tS2Rq7b5gxUdQwQ   
10888  zn81QpflLDUaGZkCMUowCg  I1rvqU2k5UQGo2lGdY6hyw   
10889  zn81QpflLDUaGZkCMUowCg  NH67MdKaFGNcP-dlu56pyw   
10890  zn81QpflLDUaGZkCMUowCg  qyNtVViurIcChc35mfYIEw   
10891  zn81QpflLDUaGZkCMUowCg  vtQOervVVTXjhvSZQiZ6PA   
10892  znDOmt2ifMXWiAkrhjiuig  AsX-6ECbV83zGJLUVMre9w   
10893  zp-DF3qfvOn5ko_vjpQLOg  KPoTixdjoJxSqRSEApSAGg   
10894  zqHznU4iL06NziZIEGWHJw  lliksv-tglfUz1T3B3vgvA   
10895  zrBmkDDLS94GYexyp0LyqQ  NH67MdKaFGNcP-dlu56pyw   
10896  zrO1ENicvYdPsQk8ykJOkg  a1t31qMLd5fQocEjbSJ61A   
10897  zs1msKnmTFD3iV2u69USuA  PXmR1MgOAWB066XH20HjxQ   
10898  zs6wQGh1r726ZzaNKRa-bw  JiLK9QPjd53pOBEAaY83lw   
10899  zs6wQGh1r726ZzaNKRa-bw  j2a5uJz76rK9uTRgLn5TdQ   
10900  zs6wQGh1r726ZzaNKRa-bw  xcxkEmy4CD-qaJUqprvpHA   
10901  zssolmdLpaX1tpRRMDWYwg  0udEgNqy5rLR5pZ4kD19Og   
10902  zulNp3NWnv7sYODNZ1Xrow  o15PeOAUzpcCl8ngk0lMHw   
10903  zuoeE7GdXXlCgr995ImWfQ  xcxkEmy4CD-qaJUqprvpHA   
10904  zv4i7JjhI9v9j4ZzX7TGDw  NCtzWkMbE13r2M2Sg0wH9w   
10905  zvFDYEFo_xO8VqLQfmB-DA  shCdCHRbnY5FTMJbWl-myQ   
10906  zw5NmE_epbvJ22xOYLdIoQ  cBwc3dhdHw0emmg9nd5SXw   
10907  zxQaAt4awDFVWme2I9mFgg  0udEgNqy5rLR5pZ4kD19Og   
10908  zxcrlC3cmH5S2TGIxuLwBw  MDtjD14H1sGLc4tSg0sUhw   
10909  zxcrlC3cmH5S2TGIxuLwBw  kGEW4XXJQ2FS94gZv_N7VA   
10910  zyor9BbfHNjTsaRFfePRwQ  apGVTRZRCQ9-89hu2qW-vw   

                                                    text  YRsentiment  \
0      I'm a big fan of this place and have dropped i...       43.539   
1      Food was ok but the service was less than exce...       -9.641   
2      I got the Penang curry and have to say the foo...        0.105   
3      To sum up in one sentence: "I only go to Thai ...      -16.096   
4      I LOVE THIS PLACE!\r\r\n\r\r\nIt's a cute mom-...        5.802   
5      The food was delicious and the service was

#### Convert Yelp Review Sentiment Scores to a 1-to-5 Scale with 0.5 Increments

In [11]:
OldMax = max(ThaiTextByUniqueUserBiz['YRsentiment'])
OldMin = min(ThaiTextByUniqueUserBiz['YRsentiment'])
NewMax = 5
NewMin = 1
OldRange = (OldMax - OldMin)
NewRange = (NewMax - NewMin)
for row in range(len(ThaiTextByUniqueUserBiz)):
    n = ThaiTextByUniqueUserBiz.loc[row, 'YRsentiment']
    ThaiTextByUniqueUserBiz.loc[row,"YRSentScore2"] = 0.5 * np.ceil(2*(((n - OldMin) * NewRange / OldRange) + NewMin))
ThaiTextByUniqueUserBiz  

user_id             business_id  \
0      --65q1FpAL_UQtVZ2PTGew  JiLK9QPjd53pOBEAaY83lw   
1      --ijvARuRJhZrBdS9_jF2A  ApUCpJ9aa6yVgsde16gYrg   
2      --ohLoec6PU9_yxhbIlVWg  2XXwiASSS6685OhWWnIt_A   
3      --qEXbk-cA0HmbPyhcffdA  CVos739DJ06t8-dNiRMyeQ   
4      --qEXbk-cA0HmbPyhcffdA  jQST5lkLGX9L52-A10TGTQ   
5      -0fMBkX7QvWKQrtOp7H-GQ  3rqoxOasrRKxNubxjLSElA   
6      -2EuoueswhqEERWezJY8gw  cInzGnaFZ3EIItvFXl1MvQ   
7      -2Ig3GSBkj8JQT8eETmDPg  d-YNxMKL6ZhkiRhfUPxKHg   
8      -3WzrbWjnaKg2QWAsouy_g  jQST5lkLGX9L52-A10TGTQ   
9      -45GJdo8Ye8A1AStuUZp9Q  -SNpLwJNup8N96yq7sBJyw   
10     -4c_mgQdLH5axJ3j2In5_Q  WPmamMTGAmNYXGoXW1mWyQ   
11     -7R1u0HzHKmhLy9qE2MBpw  KTF-E3NfkJy2wiwcgOPyVQ   
12     -8BqfYouq3o_UoazAQWwNw  a_wK-2KhPu-8DAwwRObr8A   
13     -8BqfYouq3o_UoazAQWwNw  jQST5lkLGX9L52-A10TGTQ   
14     -8gRkiYaVm3zfoQ4pcg75w  UxiSHVZxMdey7vRwm1fQyA   
15     -8pbvWZH7Czk9YW1UkW4Ng  MDtjD14H1sGLc4tSg0sUhw   
16     -9-fkZ72_Qg4E6YYYXMqSg  4nnMgD9X62YrMqkQKhx-Pg   
17     -9g6w1xoj6-4iZH29P3h7g  kGEW4XXJQ2FS94gZv_N7VA   
18     -A01aSKVuOm42FnhvOCdKA  JiLK9QPjd53pOBEAaY83lw   
19     -ARd7byPUILfnFVlKcn0Yg  wct7rZKyZqZftzmAU-vhWQ   
20     -BQFGG_hrORLkEs8oigCjg  qcm7pfIdNn9XBuPEtoogbw   
21     -BVv1TDLLphHzgKw-eAJJQ  fJzKYljToXOauSohw9cMIA   
22     -BVv1TDLLphHzgKw-eAJJQ  puFrm8eNizztqaWr_e32pQ   
23     -BrtOvg4tL1xcaQnQTaZow  oXQmAzFj_qKNhUGYGNWLSA   
24     -CyBQG3dc4UnpluY7UdMOA  ujLZmyy11g1JHCQTxRA3Dw   
25     -DRza4wuHHWfQx5HcG6qaw  NCtzWkMbE13r2M2Sg0wH9w   
26     -E7e4sTuVAHwwWjQYBG07w  5W48_DnrXVD7EbtmE4pxOQ   
27     -EFuxDYchSSVkb4Q9Iivpg  shCdCHRbnY5FTMJbWl-myQ   
28     -EyEj5BujVFisco6OwmR8A  puFrm8eNizztqaWr_e32pQ   
29     -F32Vl8Rk4dwsmk0f2wRIw  NH67MdKaFGNcP-dlu56pyw   
...                       ...                     ...   
10881  zkhOTlhe6dn-jrwDpYDN6Q  LbBxrQJl-ny02-eCM1LYNg   
10882  zlcHQII8dyI8I0LHGj8nOA  90AXjqb4O-wrTHDKDoDUzg   
10883  zlcHQII8dyI8I0LHGj8nOA  AsX-6ECbV83zGJLUVMre9w   
10884  zlneJ82kppmQXOUGHqCLaQ  RlfX4muX5LfJsvmI9qWGvw   
10885  zmZtT1T6-J4NcqP8j1L5jA  Nz_AasmpsQ8MLSqhCTRVoA   
10886  zmZtT1T6-J4NcqP8j1L5jA  joxWCp6dgN-kTE9GMziwjA   
10887  zn81QpflLDUaGZkCMUowCg  8qrICL2tS2Rq7b5gxUdQwQ   
10888  zn81QpflLDUaGZkCMUowCg  I1rvqU2k5UQGo2lGdY6hyw   
10889  zn81QpflLDUaGZkCMUowCg  NH67MdKaFGNcP-dlu56pyw   
10890  zn81QpflLDUaGZkCMUowCg  qyNtVViurIcChc35mfYIEw   
10891  zn81QpflLDUaGZkCMUowCg  vtQOervVVTXjhvSZQiZ6PA   
10892  znDOmt2ifMXWiAkrhjiuig  AsX-6ECbV83zGJLUVMre9w   
10893  zp-DF3qfvOn5ko_vjpQLOg  KPoTixdjoJxSqRSEApSAGg   
10894  zqHznU4iL06NziZIEGWHJw  lliksv-tglfUz1T3B3vgvA   
10895  zrBmkDDLS94GYexyp0LyqQ  NH67MdKaFGNcP-dlu56pyw   
10896  zrO1ENicvYdPsQk8ykJOkg  a1t31qMLd5fQocEjbSJ61A   
10897  zs1msKnmTFD3iV2u69USuA  PXmR1MgOAWB066XH20HjxQ   
10898  zs6wQGh1r726ZzaNKRa-bw  JiLK9QPjd53pOBEAaY83lw   
10899  zs6wQGh1r726ZzaNKRa-bw  j2a5uJz76rK9uTRgLn5TdQ   
10900  zs6wQGh1r726ZzaNKRa-bw  xcxkEmy4CD-qaJUqprvpHA   
10901  zssolmdLpaX1tpRRMDWYwg  0udEgNqy5rLR5pZ4kD19Og   
10902  zulNp3NWnv7sYODNZ1Xrow  o15PeOAUzpcCl8ngk0lMHw   
10903  zuoeE7GdXXlCgr995ImWfQ  xcxkEmy4CD-qaJUqprvpHA   
10904  zv4i7JjhI9v9j4ZzX7TGDw  NCtzWkMbE13r2M2Sg0wH9w   
10905  zvFDYEFo_xO8VqLQfmB-DA  shCdCHRbnY5FTMJbWl-myQ   
10906  zw5NmE_epbvJ22xOYLdIoQ  cBwc3dhdHw0emmg9nd5SXw   
10907  zxQaAt4awDFVWme2I9mFgg  0udEgNqy5rLR5pZ4kD19Og   
10908  zxcrlC3cmH5S2TGIxuLwBw  MDtjD14H1sGLc4tSg0sUhw   
10909  zxcrlC3cmH5S2TGIxuLwBw  kGEW4XXJQ2FS94gZv_N7VA   
10910  zyor9BbfHNjTsaRFfePRwQ  apGVTRZRCQ9-89hu2qW-vw   

                                                    text  YRsentiment  \
0      I'm a big fan of this place and have dropped i...       43.539   
1      Food was ok but the service was less than exce...       -9.641   
2      I got the Penang curry and have to say the foo...        0.105   
3      To sum up in one sentence: "I only go to Thai ...      -16.096   
4      I LOVE THIS PLACE!\r\r\n\r\r\nIt's a cute mom-...        5.802   
5      The food was delicious and the service was

#### Extract Sentiment Scores from Reviews Using Algorithm, TweetTokenizer and AFINN Dictionary

In [12]:
from nltk.tokenize import TweetTokenizer
tknzr = TweetTokenizer()
for row in range(len(ThaiTextByUniqueUserBiz)):
    n = ThaiTextByUniqueUserBiz.loc[row, 'text']
    words = tknzr.tokenize(n.lower())
    ThaiTextByUniqueUserBiz.loc[row,"AFINNSentScore"] = sum(sentiment_dictionary.get(word, 0) for word in words)
ThaiTextByUniqueUserBiz

user_id             business_id  \
0      --65q1FpAL_UQtVZ2PTGew  JiLK9QPjd53pOBEAaY83lw   
1      --ijvARuRJhZrBdS9_jF2A  ApUCpJ9aa6yVgsde16gYrg   
2      --ohLoec6PU9_yxhbIlVWg  2XXwiASSS6685OhWWnIt_A   
3      --qEXbk-cA0HmbPyhcffdA  CVos739DJ06t8-dNiRMyeQ   
4      --qEXbk-cA0HmbPyhcffdA  jQST5lkLGX9L52-A10TGTQ   
5      -0fMBkX7QvWKQrtOp7H-GQ  3rqoxOasrRKxNubxjLSElA   
6      -2EuoueswhqEERWezJY8gw  cInzGnaFZ3EIItvFXl1MvQ   
7      -2Ig3GSBkj8JQT8eETmDPg  d-YNxMKL6ZhkiRhfUPxKHg   
8      -3WzrbWjnaKg2QWAsouy_g  jQST5lkLGX9L52-A10TGTQ   
9      -45GJdo8Ye8A1AStuUZp9Q  -SNpLwJNup8N96yq7sBJyw   
10     -4c_mgQdLH5axJ3j2In5_Q  WPmamMTGAmNYXGoXW1mWyQ   
11     -7R1u0HzHKmhLy9qE2MBpw  KTF-E3NfkJy2wiwcgOPyVQ   
12     -8BqfYouq3o_UoazAQWwNw  a_wK-2KhPu-8DAwwRObr8A   
13     -8BqfYouq3o_UoazAQWwNw  jQST5lkLGX9L52-A10TGTQ   
14     -8gRkiYaVm3zfoQ4pcg75w  UxiSHVZxMdey7vRwm1fQyA   
15     -8pbvWZH7Czk9YW1UkW4Ng  MDtjD14H1sGLc4tSg0sUhw   
16     -9-fkZ72_Qg4E6YYYXMqSg  4nnMgD9X62YrMqkQKhx-Pg   
17     -9g6w1xoj6-4iZH29P3h7g  kGEW4XXJQ2FS94gZv_N7VA   
18     -A01aSKVuOm42FnhvOCdKA  JiLK9QPjd53pOBEAaY83lw   
19     -ARd7byPUILfnFVlKcn0Yg  wct7rZKyZqZftzmAU-vhWQ   
20     -BQFGG_hrORLkEs8oigCjg  qcm7pfIdNn9XBuPEtoogbw   
21     -BVv1TDLLphHzgKw-eAJJQ  fJzKYljToXOauSohw9cMIA   
22     -BVv1TDLLphHzgKw-eAJJQ  puFrm8eNizztqaWr_e32pQ   
23     -BrtOvg4tL1xcaQnQTaZow  oXQmAzFj_qKNhUGYGNWLSA   
24     -CyBQG3dc4UnpluY7UdMOA  ujLZmyy11g1JHCQTxRA3Dw   
25     -DRza4wuHHWfQx5HcG6qaw  NCtzWkMbE13r2M2Sg0wH9w   
26     -E7e4sTuVAHwwWjQYBG07w  5W48_DnrXVD7EbtmE4pxOQ   
27     -EFuxDYchSSVkb4Q9Iivpg  shCdCHRbnY5FTMJbWl-myQ   
28     -EyEj5BujVFisco6OwmR8A  puFrm8eNizztqaWr_e32pQ   
29     -F32Vl8Rk4dwsmk0f2wRIw  NH67MdKaFGNcP-dlu56pyw   
...                       ...                     ...   
10881  zkhOTlhe6dn-jrwDpYDN6Q  LbBxrQJl-ny02-eCM1LYNg   
10882  zlcHQII8dyI8I0LHGj8nOA  90AXjqb4O-wrTHDKDoDUzg   
10883  zlcHQII8dyI8I0LHGj8nOA  AsX-6ECbV83zGJLUVMre9w   
10884  zlneJ82kppmQXOUGHqCLaQ  RlfX4muX5LfJsvmI9qWGvw   
10885  zmZtT1T6-J4NcqP8j1L5jA  Nz_AasmpsQ8MLSqhCTRVoA   
10886  zmZtT1T6-J4NcqP8j1L5jA  joxWCp6dgN-kTE9GMziwjA   
10887  zn81QpflLDUaGZkCMUowCg  8qrICL2tS2Rq7b5gxUdQwQ   
10888  zn81QpflLDUaGZkCMUowCg  I1rvqU2k5UQGo2lGdY6hyw   
10889  zn81QpflLDUaGZkCMUowCg  NH67MdKaFGNcP-dlu56pyw   
10890  zn81QpflLDUaGZkCMUowCg  qyNtVViurIcChc35mfYIEw   
10891  zn81QpflLDUaGZkCMUowCg  vtQOervVVTXjhvSZQiZ6PA   
10892  znDOmt2ifMXWiAkrhjiuig  AsX-6ECbV83zGJLUVMre9w   
10893  zp-DF3qfvOn5ko_vjpQLOg  KPoTixdjoJxSqRSEApSAGg   
10894  zqHznU4iL06NziZIEGWHJw  lliksv-tglfUz1T3B3vgvA   
10895  zrBmkDDLS94GYexyp0LyqQ  NH67MdKaFGNcP-dlu56pyw   
10896  zrO1ENicvYdPsQk8ykJOkg  a1t31qMLd5fQocEjbSJ61A   
10897  zs1msKnmTFD3iV2u69USuA  PXmR1MgOAWB066XH20HjxQ   
10898  zs6wQGh1r726ZzaNKRa-bw  JiLK9QPjd53pOBEAaY83lw   
10899  zs6wQGh1r726ZzaNKRa-bw  j2a5uJz76rK9uTRgLn5TdQ   
10900  zs6wQGh1r726ZzaNKRa-bw  xcxkEmy4CD-qaJUqprvpHA   
10901  zssolmdLpaX1tpRRMDWYwg  0udEgNqy5rLR5pZ4kD19Og   
10902  zulNp3NWnv7sYODNZ1Xrow  o15PeOAUzpcCl8ngk0lMHw   
10903  zuoeE7GdXXlCgr995ImWfQ  xcxkEmy4CD-qaJUqprvpHA   
10904  zv4i7JjhI9v9j4ZzX7TGDw  NCtzWkMbE13r2M2Sg0wH9w   
10905  zvFDYEFo_xO8VqLQfmB-DA  shCdCHRbnY5FTMJbWl-myQ   
10906  zw5NmE_epbvJ22xOYLdIoQ  cBwc3dhdHw0emmg9nd5SXw   
10907  zxQaAt4awDFVWme2I9mFgg  0udEgNqy5rLR5pZ4kD19Og   
10908  zxcrlC3cmH5S2TGIxuLwBw  MDtjD14H1sGLc4tSg0sUhw   
10909  zxcrlC3cmH5S2TGIxuLwBw  kGEW4XXJQ2FS94gZv_N7VA   
10910  zyor9BbfHNjTsaRFfePRwQ  apGVTRZRCQ9-89hu2qW-vw   

                                                    text  YRsentiment  \
0      I'm a big fan of this place and have dropped i...       43.539   
1      Food was ok but the service was less than exce...       -9.641   
2      I got the Penang curry and have to say the foo...        0.105   
3      To sum up in one sentence: "I only go to Thai ...      -16.096   
4      I LOVE THIS PLACE!\r\r\n\r\r\nIt's a cute mom-...        5.802   
5      The food was delicious and the service was

#### Convert AFINN Sentiment Scores to a 1-to-5 Scale with 0.5 Increments

In [13]:
OldMax = max(ThaiTextByUniqueUserBiz['AFINNSentScore'])
OldMin = min(ThaiTextByUniqueUserBiz['AFINNSentScore'])
NewMax = 5
NewMin = 1
OldRange = (OldMax - OldMin)
NewRange = (NewMax - NewMin)
for row in range(len(ThaiTextByUniqueUserBiz)):
    n = ThaiTextByUniqueUserBiz.loc[row, 'AFINNSentScore']
    ThaiTextByUniqueUserBiz.loc[row,"AFINNSentScore2"] = 0.5 * np.ceil(2*(((n - OldMin) * NewRange / OldRange) + NewMin))
ThaiTextByUniqueUserBiz  

user_id             business_id  \
0      --65q1FpAL_UQtVZ2PTGew  JiLK9QPjd53pOBEAaY83lw   
1      --ijvARuRJhZrBdS9_jF2A  ApUCpJ9aa6yVgsde16gYrg   
2      --ohLoec6PU9_yxhbIlVWg  2XXwiASSS6685OhWWnIt_A   
3      --qEXbk-cA0HmbPyhcffdA  CVos739DJ06t8-dNiRMyeQ   
4      --qEXbk-cA0HmbPyhcffdA  jQST5lkLGX9L52-A10TGTQ   
5      -0fMBkX7QvWKQrtOp7H-GQ  3rqoxOasrRKxNubxjLSElA   
6      -2EuoueswhqEERWezJY8gw  cInzGnaFZ3EIItvFXl1MvQ   
7      -2Ig3GSBkj8JQT8eETmDPg  d-YNxMKL6ZhkiRhfUPxKHg   
8      -3WzrbWjnaKg2QWAsouy_g  jQST5lkLGX9L52-A10TGTQ   
9      -45GJdo8Ye8A1AStuUZp9Q  -SNpLwJNup8N96yq7sBJyw   
10     -4c_mgQdLH5axJ3j2In5_Q  WPmamMTGAmNYXGoXW1mWyQ   
11     -7R1u0HzHKmhLy9qE2MBpw  KTF-E3NfkJy2wiwcgOPyVQ   
12     -8BqfYouq3o_UoazAQWwNw  a_wK-2KhPu-8DAwwRObr8A   
13     -8BqfYouq3o_UoazAQWwNw  jQST5lkLGX9L52-A10TGTQ   
14     -8gRkiYaVm3zfoQ4pcg75w  UxiSHVZxMdey7vRwm1fQyA   
15     -8pbvWZH7Czk9YW1UkW4Ng  MDtjD14H1sGLc4tSg0sUhw   
16     -9-fkZ72_Qg4E6YYYXMqSg  4nnMgD9X62YrMqkQKhx-Pg   
17     -9g6w1xoj6-4iZH29P3h7g  kGEW4XXJQ2FS94gZv_N7VA   
18     -A01aSKVuOm42FnhvOCdKA  JiLK9QPjd53pOBEAaY83lw   
19     -ARd7byPUILfnFVlKcn0Yg  wct7rZKyZqZftzmAU-vhWQ   
20     -BQFGG_hrORLkEs8oigCjg  qcm7pfIdNn9XBuPEtoogbw   
21     -BVv1TDLLphHzgKw-eAJJQ  fJzKYljToXOauSohw9cMIA   
22     -BVv1TDLLphHzgKw-eAJJQ  puFrm8eNizztqaWr_e32pQ   
23     -BrtOvg4tL1xcaQnQTaZow  oXQmAzFj_qKNhUGYGNWLSA   
24     -CyBQG3dc4UnpluY7UdMOA  ujLZmyy11g1JHCQTxRA3Dw   
25     -DRza4wuHHWfQx5HcG6qaw  NCtzWkMbE13r2M2Sg0wH9w   
26     -E7e4sTuVAHwwWjQYBG07w  5W48_DnrXVD7EbtmE4pxOQ   
27     -EFuxDYchSSVkb4Q9Iivpg  shCdCHRbnY5FTMJbWl-myQ   
28     -EyEj5BujVFisco6OwmR8A  puFrm8eNizztqaWr_e32pQ   
29     -F32Vl8Rk4dwsmk0f2wRIw  NH67MdKaFGNcP-dlu56pyw   
...                       ...                     ...   
10881  zkhOTlhe6dn-jrwDpYDN6Q  LbBxrQJl-ny02-eCM1LYNg   
10882  zlcHQII8dyI8I0LHGj8nOA  90AXjqb4O-wrTHDKDoDUzg   
10883  zlcHQII8dyI8I0LHGj8nOA  AsX-6ECbV83zGJLUVMre9w   
10884  zlneJ82kppmQXOUGHqCLaQ  RlfX4muX5LfJsvmI9qWGvw   
10885  zmZtT1T6-J4NcqP8j1L5jA  Nz_AasmpsQ8MLSqhCTRVoA   
10886  zmZtT1T6-J4NcqP8j1L5jA  joxWCp6dgN-kTE9GMziwjA   
10887  zn81QpflLDUaGZkCMUowCg  8qrICL2tS2Rq7b5gxUdQwQ   
10888  zn81QpflLDUaGZkCMUowCg  I1rvqU2k5UQGo2lGdY6hyw   
10889  zn81QpflLDUaGZkCMUowCg  NH67MdKaFGNcP-dlu56pyw   
10890  zn81QpflLDUaGZkCMUowCg  qyNtVViurIcChc35mfYIEw   
10891  zn81QpflLDUaGZkCMUowCg  vtQOervVVTXjhvSZQiZ6PA   
10892  znDOmt2ifMXWiAkrhjiuig  AsX-6ECbV83zGJLUVMre9w   
10893  zp-DF3qfvOn5ko_vjpQLOg  KPoTixdjoJxSqRSEApSAGg   
10894  zqHznU4iL06NziZIEGWHJw  lliksv-tglfUz1T3B3vgvA   
10895  zrBmkDDLS94GYexyp0LyqQ  NH67MdKaFGNcP-dlu56pyw   
10896  zrO1ENicvYdPsQk8ykJOkg  a1t31qMLd5fQocEjbSJ61A   
10897  zs1msKnmTFD3iV2u69USuA  PXmR1MgOAWB066XH20HjxQ   
10898  zs6wQGh1r726ZzaNKRa-bw  JiLK9QPjd53pOBEAaY83lw   
10899  zs6wQGh1r726ZzaNKRa-bw  j2a5uJz76rK9uTRgLn5TdQ   
10900  zs6wQGh1r726ZzaNKRa-bw  xcxkEmy4CD-qaJUqprvpHA   
10901  zssolmdLpaX1tpRRMDWYwg  0udEgNqy5rLR5pZ4kD19Og   
10902  zulNp3NWnv7sYODNZ1Xrow  o15PeOAUzpcCl8ngk0lMHw   
10903  zuoeE7GdXXlCgr995ImWfQ  xcxkEmy4CD-qaJUqprvpHA   
10904  zv4i7JjhI9v9j4ZzX7TGDw  NCtzWkMbE13r2M2Sg0wH9w   
10905  zvFDYEFo_xO8VqLQfmB-DA  shCdCHRbnY5FTMJbWl-myQ   
10906  zw5NmE_epbvJ22xOYLdIoQ  cBwc3dhdHw0emmg9nd5SXw   
10907  zxQaAt4awDFVWme2I9mFgg  0udEgNqy5rLR5pZ4kD19Og   
10908  zxcrlC3cmH5S2TGIxuLwBw  MDtjD14H1sGLc4tSg0sUhw   
10909  zxcrlC3cmH5S2TGIxuLwBw  kGEW4XXJQ2FS94gZv_N7VA   
10910  zyor9BbfHNjTsaRFfePRwQ  apGVTRZRCQ9-89hu2qW-vw   

                                                    text  YRsentiment  \
0      I'm a big fan of this place and have dropped i...       43.539   
1      Food was ok but the service was less than exce...       -9.641   
2      I got the Penang curry and have to say the foo...        0.105   
3      To sum up in one sentence: "I only go to Thai ...      -16.096   
4      I LOVE THIS PLACE!\r\r\n\r\r\nIt's a cute mom-...        5.802   
5      The food was delicious and the service was

### Read and Merge the Thai Review Ratings file to the Sentiment Score Dataframe (ThaiTextByUniqueUserBiz)

#### Read the Thai Review Ratings File

In [14]:
ThaiReviewRatingsByUserBiz = pd.read_pickle('ThaiReviewRatingsByUserBiz.pkl')
ThaiReviewRatingsByUserBiz

user_id             business_id  review_ratings
0      --65q1FpAL_UQtVZ2PTGew  JiLK9QPjd53pOBEAaY83lw               5
1      --ijvARuRJhZrBdS9_jF2A  ApUCpJ9aa6yVgsde16gYrg               1
2      --ohLoec6PU9_yxhbIlVWg  2XXwiASSS6685OhWWnIt_A               3
3      --qEXbk-cA0HmbPyhcffdA  CVos739DJ06t8-dNiRMyeQ               3
4      --qEXbk-cA0HmbPyhcffdA  jQST5lkLGX9L52-A10TGTQ               5
5      -0fMBkX7QvWKQrtOp7H-GQ  3rqoxOasrRKxNubxjLSElA               4
6      -2EuoueswhqEERWezJY8gw  cInzGnaFZ3EIItvFXl1MvQ               4
7      -2Ig3GSBkj8JQT8eETmDPg  d-YNxMKL6ZhkiRhfUPxKHg               3
8      -3WzrbWjnaKg2QWAsouy_g  jQST5lkLGX9L52-A10TGTQ               5
9      -45GJdo8Ye8A1AStuUZp9Q  -SNpLwJNup8N96yq7sBJyw               5
10     -4c_mgQdLH5axJ3j2In5_Q  WPmamMTGAmNYXGoXW1mWyQ               5
11     -7R1u0HzHKmhLy9qE2MBpw  KTF-E3NfkJy2wiwcgOPyVQ               4
12     -8BqfYouq3o_UoazAQWwNw  a_wK-2KhPu-8DAwwRObr8A               5
13     -8BqfYouq3o_UoazAQWwNw  jQST5lkLGX9L52-A10TGTQ               5
14     -8gRkiYaVm3zfoQ4pcg75w  UxiSHVZxMdey7vRwm1fQyA               5
15     -8pbvWZH7Czk9YW1UkW4Ng  MDtjD14H1sGLc4tSg0sUhw               3
16     -9-fkZ72_Qg4E6YYYXMqSg  4nnMgD9X62YrMqkQKhx-Pg               2
17     -9g6w1xoj6-4iZH29P3h7g  kGEW4XXJQ2FS94gZv_N7VA               5
18     -A01aSKVuOm42FnhvOCdKA  JiLK9QPjd53pOBEAaY83lw               5
19     -ARd7byPUILfnFVlKcn0Yg  wct7rZKyZqZftzmAU-vhWQ               1
20     -BQFGG_hrORLkEs8oigCjg  qcm7pfIdNn9XBuPEtoogbw               5
21     -BVv1TDLLphHzgKw-eAJJQ  fJzKYljToXOauSohw9cMIA               4
22     -BVv1TDLLphHzgKw-eAJJQ  puFrm8eNizztqaWr_e32pQ               4
23     -BrtOvg4tL1xcaQnQTaZow  oXQmAzFj_qKNhUGYGNWLSA               1
24     -CyBQG3dc4UnpluY7UdMOA  ujLZmyy11g1JHCQTxRA3Dw               3
25     -DRza4wuHHWfQx5HcG6qaw  NCtzWkMbE13r2M2Sg0wH9w               4
26     -E7e4sTuVAHwwWjQYBG07w  5W48_DnrXVD7EbtmE4pxOQ               4
27     -EFuxDYchSSVkb4Q9Iivpg  shCdCHRbnY5FTMJbWl-myQ               5
28     -EyEj5BujVFisco6OwmR8A  puFrm8eNizztqaWr_e32pQ               4
29     -F32Vl8Rk4dwsmk0f2wRIw  NH67MdKaFGNcP-dlu56pyw               3
...                       ...                     ...             ...
10881  zkhOTlhe6dn-jrwDpYDN6Q  LbBxrQJl-ny02-eCM1LYNg               4
10882  zlcHQII8dyI8I0LHGj8nOA  90AXjqb4O-wrTHDKDoDUzg               4
10883  zlcHQII8dyI8I0LHGj8nOA  AsX-6ECbV83zGJLUVMre9w               2
10884  zlneJ82kppmQXOUGHqCLaQ  RlfX4muX5LfJsvmI9qWGvw               5
10885  zmZtT1T6-J4NcqP8j1L5jA  Nz_AasmpsQ8MLSqhCTRVoA               3
10886  zmZtT1T6-J4NcqP8j1L5jA  joxWCp6dgN-kTE9GMziwjA               4
10887  zn81QpflLDUaGZkCMUowCg  8qrICL2tS2Rq7b5gxUdQwQ               2
10888  zn81QpflLDUaGZkCMUowCg  I1rvqU2k5UQGo2lGdY6hyw               5
10889  zn81QpflLDUaGZkCMUowCg  NH67MdKaFGNcP-dlu56pyw               5
10890  zn81QpflLDUaGZkCMUowCg  qyNtVViurIcChc35mfYIEw               1
10891  zn81QpflLDUaGZkCMUowCg  vtQOervVVTXjhvSZQiZ6PA               3
10892  znDOmt2ifMXWiAkrhjiuig  AsX-6ECbV83zGJLUVMre9w               5
10893  zp-DF3qfvOn5ko_vjpQLOg  KPoTixdjoJxSqRSEApSAGg               5
10894  zqHznU4iL06NziZIEGWHJw  lliksv-tglfUz1T3B3vgvA               4
10895  zrBmkDDLS94GYexyp0LyqQ  NH67MdKaFGNcP-dlu56pyw               5
10896  zrO1ENicvYdPsQk8ykJOkg  a1t31qMLd5fQocEjbSJ61A               1
10897  zs1msKnmTFD3iV2u69USuA  PXmR1MgOAWB066XH20HjxQ               4
10898  zs6wQGh1r726ZzaNKRa-bw  JiLK9QPjd53pOBEAaY83lw               4
10899  zs6wQGh1r726ZzaNKRa-bw  j2a5uJz76rK9uTRgLn5TdQ               3
10900  zs6wQGh1r726ZzaNKRa-bw  xcxkEmy4CD-qaJUqprvpHA               5
10901  zssolmdLpaX1tpRRMDWYwg  0udEgNqy5rLR5pZ4kD19Og               5
10902  zulNp3NWnv7sYODNZ1Xrow  o15PeOAUzpcCl8ngk0lMHw               4
10903  zuoeE7GdXXlCgr995ImWfQ  xcxkEmy4CD-qaJUqprvpHA               1
10904  zv4i7JjhI9v9j4ZzX7TGDw  NCtzWkMbE13r2M2Sg0wH9w               5
10905  zvFDYEFo_xO8VqLQfmB-DA  shCdCHRbnY5FTMJbWl-myQ               5
10906  zw5NmE_epbvJ22xOYLdIoQ  c

#### Merge Ratings to Sentiment Scores in the ThaiTextByUniqueUserBiz Dataframe

In [15]:
ThaiTextByUniqueUserBiz = pd.merge(ThaiTextByUniqueUserBiz, ThaiReviewRatingsByUserBiz, how='left', on=['user_id', 'business_id'])
ThaiTextByUniqueUserBiz

user_id             business_id  \
0      --65q1FpAL_UQtVZ2PTGew  JiLK9QPjd53pOBEAaY83lw   
1      --ijvARuRJhZrBdS9_jF2A  ApUCpJ9aa6yVgsde16gYrg   
2      --ohLoec6PU9_yxhbIlVWg  2XXwiASSS6685OhWWnIt_A   
3      --qEXbk-cA0HmbPyhcffdA  CVos739DJ06t8-dNiRMyeQ   
4      --qEXbk-cA0HmbPyhcffdA  jQST5lkLGX9L52-A10TGTQ   
5      -0fMBkX7QvWKQrtOp7H-GQ  3rqoxOasrRKxNubxjLSElA   
6      -2EuoueswhqEERWezJY8gw  cInzGnaFZ3EIItvFXl1MvQ   
7      -2Ig3GSBkj8JQT8eETmDPg  d-YNxMKL6ZhkiRhfUPxKHg   
8      -3WzrbWjnaKg2QWAsouy_g  jQST5lkLGX9L52-A10TGTQ   
9      -45GJdo8Ye8A1AStuUZp9Q  -SNpLwJNup8N96yq7sBJyw   
10     -4c_mgQdLH5axJ3j2In5_Q  WPmamMTGAmNYXGoXW1mWyQ   
11     -7R1u0HzHKmhLy9qE2MBpw  KTF-E3NfkJy2wiwcgOPyVQ   
12     -8BqfYouq3o_UoazAQWwNw  a_wK-2KhPu-8DAwwRObr8A   
13     -8BqfYouq3o_UoazAQWwNw  jQST5lkLGX9L52-A10TGTQ   
14     -8gRkiYaVm3zfoQ4pcg75w  UxiSHVZxMdey7vRwm1fQyA   
15     -8pbvWZH7Czk9YW1UkW4Ng  MDtjD14H1sGLc4tSg0sUhw   
16     -9-fkZ72_Qg4E6YYYXMqSg  4nnMgD9X62YrMqkQKhx-Pg   
17     -9g6w1xoj6-4iZH29P3h7g  kGEW4XXJQ2FS94gZv_N7VA   
18     -A01aSKVuOm42FnhvOCdKA  JiLK9QPjd53pOBEAaY83lw   
19     -ARd7byPUILfnFVlKcn0Yg  wct7rZKyZqZftzmAU-vhWQ   
20     -BQFGG_hrORLkEs8oigCjg  qcm7pfIdNn9XBuPEtoogbw   
21     -BVv1TDLLphHzgKw-eAJJQ  fJzKYljToXOauSohw9cMIA   
22     -BVv1TDLLphHzgKw-eAJJQ  puFrm8eNizztqaWr_e32pQ   
23     -BrtOvg4tL1xcaQnQTaZow  oXQmAzFj_qKNhUGYGNWLSA   
24     -CyBQG3dc4UnpluY7UdMOA  ujLZmyy11g1JHCQTxRA3Dw   
25     -DRza4wuHHWfQx5HcG6qaw  NCtzWkMbE13r2M2Sg0wH9w   
26     -E7e4sTuVAHwwWjQYBG07w  5W48_DnrXVD7EbtmE4pxOQ   
27     -EFuxDYchSSVkb4Q9Iivpg  shCdCHRbnY5FTMJbWl-myQ   
28     -EyEj5BujVFisco6OwmR8A  puFrm8eNizztqaWr_e32pQ   
29     -F32Vl8Rk4dwsmk0f2wRIw  NH67MdKaFGNcP-dlu56pyw   
...                       ...                     ...   
10881  zkhOTlhe6dn-jrwDpYDN6Q  LbBxrQJl-ny02-eCM1LYNg   
10882  zlcHQII8dyI8I0LHGj8nOA  90AXjqb4O-wrTHDKDoDUzg   
10883  zlcHQII8dyI8I0LHGj8nOA  AsX-6ECbV83zGJLUVMre9w   
10884  zlneJ82kppmQXOUGHqCLaQ  RlfX4muX5LfJsvmI9qWGvw   
10885  zmZtT1T6-J4NcqP8j1L5jA  Nz_AasmpsQ8MLSqhCTRVoA   
10886  zmZtT1T6-J4NcqP8j1L5jA  joxWCp6dgN-kTE9GMziwjA   
10887  zn81QpflLDUaGZkCMUowCg  8qrICL2tS2Rq7b5gxUdQwQ   
10888  zn81QpflLDUaGZkCMUowCg  I1rvqU2k5UQGo2lGdY6hyw   
10889  zn81QpflLDUaGZkCMUowCg  NH67MdKaFGNcP-dlu56pyw   
10890  zn81QpflLDUaGZkCMUowCg  qyNtVViurIcChc35mfYIEw   
10891  zn81QpflLDUaGZkCMUowCg  vtQOervVVTXjhvSZQiZ6PA   
10892  znDOmt2ifMXWiAkrhjiuig  AsX-6ECbV83zGJLUVMre9w   
10893  zp-DF3qfvOn5ko_vjpQLOg  KPoTixdjoJxSqRSEApSAGg   
10894  zqHznU4iL06NziZIEGWHJw  lliksv-tglfUz1T3B3vgvA   
10895  zrBmkDDLS94GYexyp0LyqQ  NH67MdKaFGNcP-dlu56pyw   
10896  zrO1ENicvYdPsQk8ykJOkg  a1t31qMLd5fQocEjbSJ61A   
10897  zs1msKnmTFD3iV2u69USuA  PXmR1MgOAWB066XH20HjxQ   
10898  zs6wQGh1r726ZzaNKRa-bw  JiLK9QPjd53pOBEAaY83lw   
10899  zs6wQGh1r726ZzaNKRa-bw  j2a5uJz76rK9uTRgLn5TdQ   
10900  zs6wQGh1r726ZzaNKRa-bw  xcxkEmy4CD-qaJUqprvpHA   
10901  zssolmdLpaX1tpRRMDWYwg  0udEgNqy5rLR5pZ4kD19Og   
10902  zulNp3NWnv7sYODNZ1Xrow  o15PeOAUzpcCl8ngk0lMHw   
10903  zuoeE7GdXXlCgr995ImWfQ  xcxkEmy4CD-qaJUqprvpHA   
10904  zv4i7JjhI9v9j4ZzX7TGDw  NCtzWkMbE13r2M2Sg0wH9w   
10905  zvFDYEFo_xO8VqLQfmB-DA  shCdCHRbnY5FTMJbWl-myQ   
10906  zw5NmE_epbvJ22xOYLdIoQ  cBwc3dhdHw0emmg9nd5SXw   
10907  zxQaAt4awDFVWme2I9mFgg  0udEgNqy5rLR5pZ4kD19Og   
10908  zxcrlC3cmH5S2TGIxuLwBw  MDtjD14H1sGLc4tSg0sUhw   
10909  zxcrlC3cmH5S2TGIxuLwBw  kGEW4XXJQ2FS94gZv_N7VA   
10910  zyor9BbfHNjTsaRFfePRwQ  apGVTRZRCQ9-89hu2qW-vw   

                                                    text  YRsentiment  \
0      I'm a big fan of this place and have dropped i...       43.539   
1      Food was ok but the service was less than exce...       -9.641   
2      I got the Penang curry and have to say the foo...        0.105   
3      To sum up in one sentence: "I only go to Thai ...      -16.096   
4      I LOVE THIS PLACE!\r\r\n\r\r\nIt's a cute mom-...        5.802   
5      The food was delicious and the service was

#### Save the Working Dataframe for Future Reference and for Recommender Activities

In [16]:
ThaiTextByUniqueUserBiz.to_pickle('ThaiTextByUniqueUserBizWithSentimentScoresRatings.pkl')

In [17]:
ThaiTextByUniqueUserBiz.to_csv('ThaiTextByUniqueUserBizWithSentimentScoresRatings.csv', encoding='utf-8')

### Preliminary Descriptive Analysis of Sentiment Scores and Ratings

In this section of the kernel, we first provide a summary of the score and rating dataframe we created in the previous section. We add to the summary a count of values per column and each their median to assess the distribution. 

#### Draw up a Summary for the Score and Rating Working Dataframe

In [18]:
ThaiTextByUniqueUserBiz.describe()

YRsentiment   YRSentScore  YRSentScore2  AFINNSentScore  \
count  10911.000000  10911.000000  10911.000000    10911.000000   
mean      13.940278      3.009898      3.332692       13.196957   
std       19.472534      0.142946      0.260143       11.963478   
min     -233.857000      1.000000      1.000000      -36.000000   
25%        4.834500      3.000000      3.000000        6.000000   
50%       12.261000      3.000000      3.500000       11.000000   
75%       21.616000      3.000000      3.500000       18.000000   
max      249.665000      5.000000      5.000000      221.000000   

       AFINNSentScore2  review_ratings  
count     10911.000000    10911.000000  
mean          2.034002        3.869112  
std           0.185616        1.175408  
min           1.000000        1.000000  
25%           2.000000        3.000000  
50%           2.000000        4.000000  
75%           2.000000        5.000000  
max           5.000000        5.000000

#### Observations 1: 

(1) The sentiment score results from the algorithm and the Yelp Restaurant Review Sentiment Lexicon range from -234 to 250.

(2) The sentiment score results from the algorithm and the two AFINN Lexicons combined range from -36 to 221.

#### Observations 2: 

(1) Like the review_ratings, both sets of converted sentiment scores range from 1 to 5. The mean changes: it is the lowest at 3.0 with sentiment scores converted with no decimals, 3.3 with sentiment scores converted to a 0.5-increment scale and 3.9 with review ratings.

#### Perform a Count of Yelp Review Sentiment Scores Scaled to 1-to-5 with No Decimals and Compute the Median

In [19]:
pd.value_counts(ThaiTextByUniqueUserBiz.YRSentScore.ravel())

3    10699
4      156
2       52
5        3
1        1
dtype: int64

In [20]:
ThaiTextByUniqueUserBiz.YRSentScore.median()

3.0

#### Observations 3:

(1) The Yelp Restaurant Review Lexicon produced a range of values where 10,699 values out of 10,911 lie in the middle which happens to be the median. This median corresponds to a count of 0 in the initial score count; in other words, the median corresponds to an evenly balanced score of sentiments or a neutral sentiment, which is neither positive nor negative. 

(2) This leaves us with 156 + 3 Thai restaurant reviews in the Yelp dataset (given reviewers of Phoenix restaurants) for recommendation purposes that have sentiments that are either positive or very positive. A pertinent question might be how many of these 159 positive-sentiment elements are reviews for Thai restaurants in Las Vegas.

#### Perform a Count of Yelp Review Sentiment Scores Scaled to 1-to-5 with 0.5 Increments and Compute the Median

In [21]:
pd.value_counts(ThaiTextByUniqueUserBiz.YRSentScore2.ravel())

3.5    6987
3.0    3712
4.0     147
2.5      48
4.5       9
2.0       4
5.0       3
1.0       1
dtype: int64

In [22]:
ThaiTextByUniqueUserBiz.YRSentScore2.median()

3.5

In [23]:
ThaiTextByUniqueUserBiz[ThaiTextByUniqueUserBiz['YRsentiment']==0]

user_id             business_id    text  YRsentiment  \
10722  ywf8LhV2jCvbksPx6wc7aw  ApUCpJ9aa6yVgsde16gYrg  Closed            0   

       YRSentScore  YRSentScore2  AFINNSentScore  AFINNSentScore2  \
10722            3             3               0                2   

       review_ratings  
10722               1

#### Observations 4:

(1) This version of the Yelp Review Sentiment Scores Scaled 1-to-5 with 0.5 increments offers new potentials for a recommender system as 6,987 + 147 + 9 + 3 reviews fare positively. 3,712 reviews are neutral whereas 48 + 4 + 1 reviews raise a negative sentiment.

(2) The distribution is skewed to positive (although not quite as dramatically as the review_ratings (see below)).

(3) As there is no 1.5 score, we acknowledge that the one review with a score of 1 may be an outlier.

(4) The filter "0" applied to the initial score count without scaling ('0' is the neutral value in this column) suggests that the neutral value for the corresponding scaled value (without decimals) is 3 and that the neutral value for the corresponding scaled value using 0.5 increments is also 3. 

(5) The median of 3.5 is different than the corresponding scaled value of 3 (using 0.5 increments) in item 4.

#### Perform a Count of AFINN Sentiment Scores Scaled to 1-to-5 with 0.5 Increments and Compute the Median

In [24]:
pd.value_counts(ThaiTextByUniqueUserBiz.AFINNSentScore2.ravel())

2.0    9573
2.5     952
1.5     328
3.0      54
3.5       2
1.0       1
5.0       1
dtype: int64

In [25]:
ThaiTextByUniqueUserBiz.AFINNSentScore2.median()

2.0

In [26]:
ThaiTextByUniqueUserBiz[ThaiTextByUniqueUserBiz['AFINNSentScore']==0]

user_id             business_id  \
172    08PXVzu6ysM93s-HaUPOIQ  90AXjqb4O-wrTHDKDoDUzg   
192    0GYLQqgu7v_qMZGgflv5Hg  4xpACaa99_KFokYvNLXMBA   
206    0ItqxKwFSTgkifR6Fv13dQ  shlAd7PLzWlQrkQ0uYcBBg   
233    0U6YfdxR2aw-_4vctk9eKA  2bdKR3l4o-S1CscLqqnvVw   
242    0WPwr3Jr4a_Ywg_hZxm4KQ  1621ir5mjVgbHwxCbMAEjg   
328    0npnrzUAhaiOX2awq3dPUw  4nnMgD9X62YrMqkQKhx-Pg   
351    0zcc8klD5N3kUvHnpA5ZlA  TqVjy0dxvNh51BF9KePCoQ   
390    1AM1mfGPGIlQi78p_OtnPQ  CVos739DJ06t8-dNiRMyeQ   
396    1CpyQ1VkgmNwA6RxHoq2bA  KPoTixdjoJxSqRSEApSAGg   
410    1J3EW_WdsDoOeQv5j1Ws8w  17DI33J8TkcfzyoiIYLQIw   
429    1Sujllo0dj27Yn1GyjLc0A  PXmR1MgOAWB066XH20HjxQ   
504    1ykGYq1ERicaZWCl5-Vi0Q  JiLK9QPjd53pOBEAaY83lw   
596    2UPKuscn_Th2C72OSn7Z8Q  KPoTixdjoJxSqRSEApSAGg   
630    2fJXLk_fkVdLAQmHLCMi-Q  1621ir5mjVgbHwxCbMAEjg   
661    2rawUU9YgDUy9mTXj5NuSw  oJpmYvLibGrYPDvcaUeMOw   
732    3EyjZYtNuzA0YY-2sDULoQ  mPGgxatANSPw9KbMluhXkA   
854    3nDDsqAEpRa_wiDEz7Y4Ww  JiLK9QPjd53pOBEAaY83lw   
885    3xmdCeTGOQLB9w2qXk5QGA  QC18oZ4atjW1vtZo71ohxw   
889    3xmdCeTGOQLB9w2qXk5QGA  uY1hOM4pySx07Yle9NGAiQ   
903    43czilYWn6VjRWskwN2_9Q  E4b5OC_6mZ0V7B6Nyjncsg   
907    441ApXa2cQHpsYtdN4Jvxw  nefBMCGgMPvOnASoyO_X1A   
916    47jZmzne8zhWEuCm8aZj8A  j2a5uJz76rK9uTRgLn5TdQ   
936    4D6fUykocgc99jXPKyYzTA  90AXjqb4O-wrTHDKDoDUzg   
938    4DX3XjoOaR9e4y3EQPoHSw  -SNpLwJNup8N96yq7sBJyw   
946    4E_nPWw89FLFHdNsEgMH-g  a1t31qMLd5fQocEjbSJ61A   
967    4GdN3itvzmFc_6kib2UzlA  2bdKR3l4o-S1CscLqqnvVw   
1005   4UUIpbOTPmu43wuC2aSGkg  puFrm8eNizztqaWr_e32pQ   
1027   4aYStmYjksIS4PPf2_l49g  km3g-wDO2KfhfnTvJrLJig   
1109   58qQ6J55I1SiCMHl4PpQ0w  SMpL3z4FLF07bRA6-y22JQ   
1136   5Ka8MMYEoZfsU-jZzUt00Q  JiLK9QPjd53pOBEAaY83lw   
...                       ...                     ...   
9500   rj98ZLijdmoq6XdW_oodJQ  4dL8yzQu974eCtTI-V4Znw   
9621   sPFh-si4S3Ii6A-1Ycg99g  ApUCpJ9aa6yVgsde16gYrg   
9647   sYGCmAr5HWknrMoihaLJ8g  UxiSHVZxMdey7vRwm1fQyA   
9752   tCVQyYHcmOrx2i81C47Tew  K4DHwck_2ds1wURnmZdFiA   
9784   tQ4FIilp-7oz4A5hRMy4EQ  NCtzWkMbE13r2M2Sg0wH9w   
9854   tlSSQwfHYJany7wPoTH46A  wct7rZKyZqZftzmAU-vhWQ   
9895   u1KWcbPMvXFEEYkZZ0Yktg  RFeDe3fNr14kvUKlVx6_4w   
9963   uQhrZB0oIJoMdpesLawWcw  -SNpLwJNup8N96yq7sBJyw   
9992   uZbTb-u-GVjTa2gtQfry5g  puFrm8eNizztqaWr_e32pQ   
9996   uaLKzVZZY8B6_u_yfER3nw  5hfQ5cNFDFPhcjx2De9-qQ   
10023  ukXbRRMYn6I22UQtQ6STdg  JiLK9QPjd53pOBEAaY83lw   
10034  uqEQsMbvsQQcqKpZrPyYwQ  vtQOervVVTXjhvSZQiZ6PA   
10169  vYmM4KTsC8ZfQBg-j5MWkw  jQST5lkLGX9L52-A10TGTQ   
10191  vhxFLqRok6r-D_aQz0s-JQ  NBvrN_ZDpmBCsPI1qLj1Qw   
10259  wBl40prH1tWmtvjtoPGcaA  SMpL3z4FLF07bRA6-y22JQ   
10310  wRyLeAg0R9WMFg3gqpnX-Q  qyNtVViurIcChc35mfYIEw   
10333  waD2euOMPTnTzQ0thq3H6Q  3ddQnTTIY-mdsBMn9b53Cw   
10346  wbhLZERyuuFSxrk441waBQ  IRxBQfA7FdHhzrPjtBxuuw   
10387  wo4b_NZrBfWLoPE0c_9Qng  oJpmYvLibGrYPDvcaUeMOw   
10495  xaXarFsZsvd3OhhutFNcNA  kaeF6dd_f_Hd7raszKWAWg   
10531  xnVrFpRkvTffPdisdDzw_A  wzuIHNOJrxSwfhKt-wV1TA   
10534  xoeDPWM2khcwMPfDoFzmKg  KTF-E3NfkJy2wiwcgOPyVQ   
10579  y3sOyQUiwL_dV38H652dfw  3ddQnTTIY-mdsBMn9b53Cw   
10607  yCZWhNAVXeZHv_g08P4x_Q  cT_rocMh92B9t62Disp6gA   
10623  yIYQewxT56mvIDLuBm7k9Q  1621ir5mjVgbHwxCbMAEjg   
10670  yfl_yB-Bv4BTgtTAVMFR6w  lliksv-tglfUz1T3B3vgvA   
10722  ywf8LhV2jCvbksPx6wc7aw  ApUCpJ9aa6yVgsde16gYrg   
10786  zHgmWv6NbwZepUSqA_FlLw  VbXy3tH5RAu7HjT7VeMMgA   
10879  zjWmhqGl_L1F8VzuuzaDUQ  mPGgxatANSPw9KbMluhXkA   
10899  zs6wQGh1r726ZzaNKRa-bw  j2a5uJz76rK9uTRgLn5TdQ   

                                                    text  YRsentiment  \
172    Very nicely appointed place, nicely done. The ...        3.434   
192    Spring roll was chewy and oily. Soup flavor pr...      -11.058   
206    Been here a couple times and it continues to s...        1.855   
233    I used to come here all the time for take out ...        1.156   
242    When you ask for HOT it means hot. Not take it...        0.166   
328    Standard wok place. My experience was that

#### Observations 5:

(1) As the median is 2 (not 3 like what the results show from the Yelp Review Sentiment Score code cells above), the distribution is skewed to positive although 9,573 out of 10,911 reviews are neutral in sentiment.

(2) As there are no 4.0 and 4.5 scores, we acknowledge that the one review with a score of 5 may be an outlier.

(3) There are more neutral reviews from the computations with the AFINN Lexicon than there are with the Yelp Restaurant Review Lexicon in a 0.5 incremental scale.

(4) Only 952 + 54 + 2 + 1 reviews raise a positive sentiment.

(5) The filter "0" applied to the initial AFINN score count without scaling ('0' is the neutral value in this column) suggests  that the neutral value for the corresponding scaled value using 0.5 increments is 2. There are 252 reviews that correspond to a neutral position according to the AFINN Lexicon initial count whereas there is one only one review that corresponds to a neutral position according to the Yelp Restaurant Review Lexicon count. The latter has also a neutral value of 2 in the scaled AFINN column. Interestingly, the corresponding rating given by the user is "1" which is "strongly dislike".  The only word in that particular review is "closed" which likely means that it is not in the lexicons or that it is rated as neutral. 

#### Perform a Count of Ratings Given by Individual Users and Compute the Median

In [27]:
pd.value_counts(ThaiTextByUniqueUserBiz.review_ratings.astype(int).ravel())

5    3949
4    3775
3    1586
2     918
1     683
dtype: int64

In [28]:
ThaiTextByUniqueUserBiz.review_ratings.astype(int).median()

4.0

#### Observations 6: 

(1) As the median is 4 (not 2 or 3 like the results from the Yelp and AFINN lexicons), the distribution is clearly skewed to "liking" with 3,949 + 3,775 reviews that are leaning to "liking", 1,586 reviews that are satisfactory, and 918 + 683 reviews that are leaning towards 'disliking'.   

(2) Although Yelp did not use 0.5 increments (we selected integer as a type to prevent function-produced decimals), the distribution offers a good range of possibilities. They do not match the range of possibilities produced by the sentiment detection methods above.

### CONCLUSIONS

The limited number of positively-charged reviews using either lexicons may improve the odds of predicting top restaurants for a user. The very high number of neutral reviews using either lexicons is a little disappointing: we hoped to see a better range of sentiment scores. We anticipate that a normalizing step may necessary for better recommendation results before the scaling phase in this sentiment detection exercise; otherwise, the high number of neutral reviews may remain when it's time to normalize the distribution in the recommender's set of activities.   